---
title: "JACUSA2helper"
author: "Michael Piechotta"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
bibliography: JACUSA2helper.bib
link-citations: yes
vignette: >
  %\VignetteIndexEntry{JACUSA2helper}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo = FALSE, message = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
options(tibble.print_min = 4L, tibble.print_max = 4L)
library(JACUSA2helper)
library(magrittr)
set.seed(777)
```

A typical workflow for analyzing [JACUSA2](https://github.com/dieterich-lab/JACUSA2) 
output file(s) with [JACUSA2helper](https://github.com/dieterich-lab/JACUSA2helper) 
consists of:

* Read JACUSA2 output file(s) into result object.
* Add optional data, e.g.: `base_sub()` - add base substitution.
* Filter result object by some criteria, e.g.: `dplyr::filter(All(cov$cond1 >= 10))` - 
  retain sites with read coverage >= 10, and use `All()` or `Any` on structured 
  columns (conditions and replicates).
* Plot filtered result object.

JACUSA2helper supports the analysis of the following methods from JACUSA2: 
*call-{1,2}*, *pileup*, *rt-arrest*, and *lrt-arrest*(experimental).
Check the [JACUSA2 manual](https://github.com/dieterich-lab/JACUSA2/blob/master/manual/manual.pdf)
for more details.

The main data structure is the result object that is implemented by following 
the tidy data approach from @Wickham2014 to feature easy interaction with 
[dplyr](https://github.com/tidyverse/dplyr) and [ggplot2](https://github.com/tidyverse/ggplot2).

In the following, we will focus on single pairwise comparisons of two conditions 
of one output file. If you want to analyze multiple files with similar pairwise 
comparisons, read `vignette("JACUSA2helper meta conditions")`. 

## Multi site and new features
While previously in JACUSA*1*, a site could be uniquely identified as one line 
in the output by the coordinate columns: 'contig', 'start', 'end', and 'strand', 
JACUSA*2* features more complex data structures to store new features such as 
arrest positions of arrest events, variant tagging, and INDEL counts. 
Therefore, sites can now cover multiple lines of output - check Section 
"Introduction" of [JACUSA2 manual](https://github.com/dieterich-lab/JACUSA2/blob/master/manual/manual.pdf) 
for details.

## Variant calling - Data: rdd (RNA-DNA-differences)
To introduce the basic verbs for manipulating JAUCSA2 output, we'll use 
`JACUSA2helper:rdd`. This data set contains a subset from @Piechotta2017 and is 
documented in `?rdd`. In brief, the sample data set consists of 10.000 sites of 
1x DNA and 2x stranded RNA sequencing libraries that allow to define 
RNA-DNA-differences (RDDs).

```{r}
library(JACUSA2helper)
data(rdd)
```

Note that `rdd` is a tibble - check [tibble](http://tibble.tidyverse.org) for 
more details. 

Use dplyr to manipulate the result object `rdd`. Check details on 
[dplyr](https://github.com/tidyverse/dplyr) and read `vignette("dplyr")`.

### Structured column
JACUSA2 features structured columns (nested tibbles) where condition and replicate 
specific data is stored, e.g.: bases, cov, etc.
```{r}
str(rdd[c("id", "bases", "cov")])
```

Check structure of column with:
```{r}
names(rdd$cov)
```
To access specific coverage information for condition 2 use:
```{r}
str(rdd$cov$cond2)
```
In summary, all structured/nested columns hold:
* data of a condition "cond*I*" on the first level, and
* data of a replicate "rep*J*" on the second level.

Coverage information for condition 2 and replicate 1 can be accessed with:
```{r}
str(rdd$cov$cond2$rep1)
```

#### Wrapper for `lapply()` and `mapply()`
For accessing and manipulating structured columns there are JACUSA2 specific 
wrappers for `lapply()` and `mapply()`:
* `lapply_cond(x, f)` operates on condition data (list of replicate data),
* `lapply_repl(x, f)` operates on replicate data of each condition, and
* `mapply_repl(f, ...)` operates on replicate data, takes multiple inputs.

To get the total coverage for each condition use:
```{r}
lapply_cond(rdd$cov, rowSums)
```

The following gives you the observed base calls:
```{r}
base_calls <- lapply_repl(rdd$bases, base_call)
base_calls$cond2
```

### Filtering with `dplyr::filter()`
dplyr and `%>%` from magrittr allow to formulate compact analysis pipelines. 

In a first step, we will retain sites with score >= 2.

```{r}
# before filtering
dim(rdd)
result <- dplyr::filter(rdd, score >= 2)
print("After filtering")
# after filtering
dim(result)
```

In order to formulate complicate statements with `dplyr::filter()` featuring 
arbitrary combinations of conditions and/or replicates use the following 
convenience functions:
* `All()` and
* `Any()`.

The following statement retains sites where all replicates of condition 1 have 
coverage >= 50 and at least one replicate of condition 2 has 10 reads:
```{r}
rdd %>% 
  dplyr::filter(All(cov$cond1 >= 50) & Any(cov$cond2 >= 10)) %>%
  dplyr::select(id, cov)
```

In the following, we will:

* remove sites with score < 2,
* retain sites with coverage >= 10 for all replicates,
* remove sites with > 2 observed bases (excluding reference base),
* apply a filter that retains only robust sites (RNA editing must be present in 
  all/both replicates - see TODO),
* Finally, we compute the base substitution (here RNA editing sites against FASTA 
  reference) per replicate.

```{r}
filtered <- rdd
filtered[["bc"]] <- lapply_cond(filtered$bases, function(b) { Reduce("+", b) } ) %>% Reduce("+", .) %>% base_count()
filtered %>%
  dplyr::filter(score >= 2) %>%
  dplyr::filter(All(cov$cond1 >= 10) & All(cov$cond2 > 10)) %>%
  dplyr::filter(bc <= 2) %>%
  dplyr::filter(robust(bases))

# sum base call counts of condition / RNA replicates
rna_bases <- Reduce("+", filtered$bases$cond2)

# we don't need lapply_repl, because we don't operate on all replicate from all
# conditions - only condition 2 / RNA
ref2rna <- base_sub(rna_bases, filtered$ref)
table(ref2rna)

```

Instead of using DNA information from the reference FASTA sequence via column "ref",
we could use the actual data from condition 1 / DNA.

TODO Robust requires observations (here base calls) to be present in all replicates of at least condition.

```{r}
cond1_ref <- base_call(filtered$bases$cond1$rep1)
dna2rna <- base_sub(rna_bases, filtered$ref)
table(dna2rna)
```

Using different sources of DNA to describe RRDs via `ref2rna` and `dna2rna` 
we can conclude that there are no polymorphic positions in `filtered` because 
`ref2rna` and `dna2rna` are identical.

### Plot base substitutions
We can now plot the site specific distribution of base substitutions using [ggplot2](https://github.com/tidyverse/ggplot2) with:

```{r}
tidyr::tibble(base_sub = dna2rna) %>% # ggplot requires a data frame
  ggplot2::ggplot(ggplot2::aes(x = base_sub, fill = base_sub)) +
  ggplot2::geom_bar() +
  ggplot2::xlab("Base substitution")  +
  ggplot2::theme(
   legend.position = "none", 
   axis.text.x = ggplot2::element_text(angle = 90, vjust = 0.5, hjust=1)
  )
```

### Plot coverage distribution

To compare the coverage distribution of each data sample, we create a simple plot of the empirical coverage distribution, where we map cond(ition) to `colour` and rep(licate) to `linetype`.

```{r}
gather_repl(result$id, result$cov) %>%
  ggplot2::ggplot(ggplot2::aes(x = value, colour = cond, linetype = repl)) + 
    ggplot2::stat_ecdf(geom = "step") +
    ggplot2::xlab("Coverage")
```


In order to improve the preliminary plot, we need to modify the following aspects:

* Merge legends,
* add descriptive plot labels (x = Coverage, y = Density),
* add descriptive data description - What is cond(ition) 1 and 2, and finally
* logarithmize cov(erage).

First, a data description is added to the result object:

```{r}
result <- gather_repl(rdd$id, rdd$cov) %>%
  dplyr::mutate(
    data_desc = dplyr::case_when(
      cond == "cond1" ~ "DNA",
      cond == "cond2" & repl == "rep1" ~ "RNA 1",
      cond == "cond2" & repl == "rep2" ~ "RNA 2"
    )
  )
```


Supplementary data `limits` and `labels` are created to enable sleek legend by relating cond(ition) and repl(icate) to data description.
They will be mapped to `colour` and `linetype`.

```{r}
# grouping enables nice legends in plot
result$group <- do.call(interaction, result[c("cond", "repl")])
# relate group={cond(ition), repl(icat)}, and nice label
meta_desc <- dplyr::distinct(result, cond, repl, group, data_desc)
# map group values to nice labels (data_desc)
limits <- as.vector(meta_desc[["group"]])
labels <- meta_desc[["data_desc"]]
```

Finally, we combine the previous snippets and arrive at the final plot:

```{r}
name <- "Data description"
result %>% 
  ggplot2::ggplot(ggplot2::aes(x = value, colour = group, linetype = group)) +
  # map values for cond to colour and use nice labels
  ggplot2::scale_colour_manual(
    name = name,
    labels = labels,
    limits = limits,
    values = factor(meta_desc[["cond"]]) %>% as.integer()
  ) +
  # map values for repl to colour and use nice labels
  ggplot2::scale_linetype_manual(
    name = name,
    labels = labels,
    limits = limits,
    values = factor(meta_desc[["repl"]]) %>% as.integer()
  ) +
  ggplot2::stat_ecdf(geom = "step") +
  # theming and xy labels
  ggplot2::ylab("Density") +
  # logarithmize and nice x-axis
  ggplot2::scale_x_log10(
    breaks = scales::trans_breaks("log10", function(x) 10^x),
    labels = scales::trans_format("log10", scales::math_format(10^.x))
  ) +
  ggplot2::xlab("Coverage") +
  ggplot2::theme(legend.position = "bottom")  
```

From the plot, we can deduce that RNA samples have similar coverage distribution 
and are higher covered than the DNA data sample. This code serves as a blueprint 
for other plots, e.g.: 

* empirical cumulative distribution of arrest rate or
* number of observed bases.

## Arrest events

In @Zhou2018 the authors map RNA modification of pseudouridine ($\Psi$) by 
chemically modifying pseudouridines with carbodiimide (+CMC) and detecting 
arrest events that are induced by reverse transcription stops in high-throughput 
sequencing under 3 different conditions: HIVRT, SIIIRTMn, and SIIIRTMg. 
The result of JACUSA2 are available by using `data()`. 
Read [JACUSA2 manual](https://github.com/dieterich-lab/JACUSA2/blob/master/manual/manual.pdf)
for details on how data has been processed from the original publication by @Zhou2018. 
In brief, the data has been filtered to contain the following rRNAs: RNA18SN5 and RNA28SN5.

In the following, we will be looking at the results for `data(HIVRT)`. 
By default JACUSA2 output from *rt-arrest* will contain the following 
structured/nested columns:

* `arrest` and `through`
* `bases` = `arrest` + `through`: total base counts.
* `arrest_rate`
* `cov` holds coverage for total base counts.

```{r}
data(HIVRT)
# column names of arrest data
names(HIVRT)
```

We investigate the strand specific coverage distribution of reads and discover that more reads are mapped to the "+" strand.
The other mappings are probably artefacts. But this discussion is beyond the scope of this vignette.

```{r}
gather_repl(HIVRT$id, HIVRT$cov) %>% # extract coverages
dplyr::inner_join(dplyr::select(HIVRT, id, contig, strand), by = "id") %>% # add contig and strand
ggplot2::ggplot(ggplot2::aes(x = value, colour = strand)) + 
  ggplot2::geom_density() + ggplot2::scale_x_log10(
    breaks = scales::trans_breaks("log10", function(x) 10^x),
    labels = scales::trans_format("log10", scales::math_format(10^.x))
  ) +
ggplot2::xlab("Coverage") +
ggplot2::theme(legend.position = "bottom") + 
ggplot2::facet_grid(contig ~ .)
```

We filter by pvalue, coverage, strand, and only retain robust arrest events:

```{r}
dim(HIVRT)
filtered <- HIVRT %>% 
  dplyr::filter(strand == "+") %>%
  dplyr::filter(pvalue <= 0.00001) %>% 
  dplyr::filter(All(cov$cond1 >= 100) & All(cov$cond2 >= 100)) %>%
  dplyr::filter(robust(arrest))
dim(filtered)
```
We use `gather_repl()` function to transform the filtered result object and 
plot the arrest rate from each condition in a scatter plot.

```{r}
filtered %>%
  ggplot2::ggplot(ggplot2::aes(x = arrest_rate$cond1$rep1, y = arrest_rate$cond2$rep1)) +  
  ggplot2::geom_point() + 
  ggplot2::facet_grid(. ~ contig) + 
  ggplot2::geom_abline(colour = "red") +
  ggplot2::xlab("+CMC arrest rate") + 
  ggplot2::ylab("-CMC arrest rate")
```

From the above plot, we can deduce that the arrest rate is higher in the +CMC condition for both rRNAs when compared against -CMC.

If you want to look simultaneously at all conditions (HIVRT, SIIIRTMn, and SIIIRTMg), checkout `vignette("JACUSA2helper-meta-conditions")`.


## Stranded data
When working with stranded RNA-Seq data, inverting base calls is not necessary because JACUSA2 will automatically invert Single End (SE) and Paired End (PE) depending on the provided library type option "-P" UNSTRANDED|FR_FIRSTSTRAND|RF_SECONDSTRAND".

# Input/Output
There are two functions to read JACUSA2 output `read_result(file)` and `read_results(files, meta_conds)` which is explained in  `vignette("JACUSA2helper-meta-conditions")`.

Use `result <- read_result("JACUSA2.out")` to read and create a JACUSA2 result object from a JACUSA2 output file `JACUSA2.out`.
By default, variables stored in the **info** column will be processed and unpacked making them available for further manipulation.  

Summary of input/output methods:

* `read_result()` Reads and unpacks a JACUSA2 result file and creates a result object.
* `read_results()` Allows to combine multiple result files and distinguish them with meta conditions.
* `write_bedGraph()` Writes a vector of values as bedGraph file.

# General function layout
In the following the core functions of JACUSA2helper are presented. Check the respective help page, e.g.: `?bc_ratio` to get more details.

`arrest_rate` Calculate arrest rate from arrest and through reads.
`base_call` Calculates observed bases from base counts. 
`base_ratio` Calculates a ratio matrix from a base count matrix.
`coverage` Calculates coverage from base counts. In most cases this will be called automatically in `read_result()`
`non_ref_ratio` Calculates non reference base ratio for base counts of some base type.
`base_sub` Calculates base substitution from reference and base counts, e.g.: A->G.
`sub_ratio` Calculates base substitution ratio from reference and observed bases, e.g.: 25% A->G.

# Filter result object
* `robust()`
* `filter_artefact()`
* `All(), Any()`

# Variant tagging
Variant tagging allows to group data based on base substitutions on reads, called tags.
Running JACUSA2 with `-B "A2G"` will tag all reads that have a "A->G" base substitution.
Where available, each observations for each site (=contig, start, stop, strand) from tagged and all reads will be provided and
distinguished by "tag=A2G" or "tag=*" in the info column, respectively.
Use `result$tag <- clean_tag(result$tag)` to convert base substitutions such as "A2G" to "A->G".

<!--
#```{r}
## load data
#data(APOBEC1)
#result <- APOBEC1
#barplot(table(result$tag), ylab="# of sites", main="Distribution of tagged sites")
#```
#We calculate the total base call count with `lapply_cond` and `Reduce("+")`.

#```{r}
## sum base call counts of conditions and replicates
#result$total_bases <- lapply_cond(
#  result$bases, function(bases) { Reduce("+", bases) } 
#) %>% Reduce("+", .)
#```


#```{r}
## mark "valid" sites 
#result <- result %>%
#  dplyr::mutate(
#    keep = (
#      # restrict to total reads
#      tag == "*" & 
#      # require less than 2 different base calls
#      base_count(total_bases, ref) <= 2 & 
#      # require 10 reads for each replicate of each condition
#      All(cov$cond1 >= 10) & All(cov$cond2 >= 10) &
#      score >= 2 &
#      robust(bases)
#    )
#  )
## raw data
## add structured columns: tagged_bases and not_tagged_bases
#expanded <- expand_tag(result)
#filtered <- dplyr::filter(expanded, keep == TRUE)
#```

#```{r}
## plot
#count <- c(nrow(result), nrow(expanded), nrow(filtered))
#b <- barplot(count, names.arg = c("raw", "expaded", "filtered"), ylim=c(0, max(count) + max(count) * 15 / 100))
#text(b, count + count * 10 / 100, count)
#```

#```{r} 
## add observed non reference base substitution, e.g.: A->G
#filtered[["base_sub"]] <- base_sub(filtered$total_bases, filtered$ref)
## ratio of non reference base calls
#filtered[["non_ref_ratio"]] <- lapply_repl(filtered$bases, function(cond) non_ref_ratio(cond, filtered$ref))
#
## tag specific base substitution for each condition
#filtered[["tagged_base_sub"]] <- lapply_cond(filtered$tagged_bases, function(bases) { Reduce("+", bases) %>% base_sub(filtered$ref)} )
#filtered[["tagged_non_ref_ratio"]] <- lapply_repl(filtered$tagged_bases, function(cond) non_ref_ratio(cond, filtered$ref))
#
#filtered[["not_tagged_base_sub"]] <- lapply_cond(filtered$not_tagged_bases, function(bases) { Reduce("+", bases) %>% #base_sub(filtered$ref)} )
#filtered[["not_tagged_non_ref_ratio"]] <- lapply_repl(filtered$not_tagged_bases, function(cond) non_ref_ratio(cond, filtered$ref))
#```
-->

<!--
# Experimental features
* *lrt-arrest*
-->

## References
