<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Auxiliary R package for the assessment of JACUSA1.x and JACUSA2.x results."><title>JACUSA2helper: A package for post-processing JACUSA2 result files. — JACUSA2helper • JACUSA2helper</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="JACUSA2helper: A package for post-processing JACUSA2 result files. — JACUSA2helper"><meta property="og:description" content="Auxiliary R package for the assessment of JACUSA1.x and JACUSA2.x results."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">JACUSA2helper</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.9.9.9500</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/JACUSA2helper.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/JACUSA2helper-dart-seq.html">Analysis of DART-seq with JACUSA2helper</a>
    <a class="dropdown-item" href="../articles/JACUSA2helper-mazf.html">Analysis of m6A mapping by MazF with JACUSA2helper</a>
    <a class="dropdown-item" href="../articles/JACUSA2helper-meta-condtions.html">Introduction to meta conditions with JACUSA2helper</a>
    <a class="dropdown-item" href="../articles/web_only/JACUSA2helper-nanopore.html">Analysis of Nanopore HEK293 with JACUSA2helper</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/dieterich-lab/JACUSA2helper/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>JACUSA2helper: A package for post-processing JACUSA2 result files.</h1>
      <small class="dont-index">Source: <a href="https://github.com/dieterich-lab/JACUSA2helper/blob/HEAD/R/common.R" class="external-link"><code>R/common.R</code></a></small>
      <div class="d-none name"><code>JACUSA2helper.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Auxiliary R package for the assessment of JACUSA1.x and JACUSA2.x results.</p>
    </div>


    <div class="section level2">
    <h2 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a></h2>
    

<p>A package that provides functions to post-process result files of JACUSA2.</p>
<p>The following methods from JACUSA2 are supported:</p><dl><dt>call1,2</dt>
<dd><p>calling variants from one condition or a comparison of conditions.</p></dd>

 <dt>pileup</dt>
<dd><p>SAMtools like pileup.</p></dd>

 <dt>rt-arrest</dt>
<dd><p>identification of transcription arrest events.</p></dd>

 <dt>lrt-arrest</dt>
<dd><p>combination of variant discovery and read arrest events.</p></dd>


</dl><p>One major new feature of JACUSA2 is the identification of read arrest events. 
In this method, the vector of base call is partitioned into read arrest and read through bases.</p>
<p>When working with stranded RNA-Seq data, inverting base calls is not necessary because
JACUSA2 will automatically invert Single End (SE) and Paired End (PE) depending on the
provided library type option "-P" UNSTRANDED|FR_FIRSTSTRAND|RF_SECONDSTRAND".</p>
<p>The central data structure in JACUSA2helper is the JACUSA2 result object that follows the 
tidy data approach to feature easy interaction with dplyr and ggplot2.
A JACUSA2 result object can be created via <code>result &lt;- read_result("jacusa2.out")</code> and is 
currently represented as a tibble. Special structured columns exist that 
hold condition andn replicate related data such as: coverage, bases. arrest rate.
Furthermore, JACUSA2helper supports the analysis of several related  JACUSA2 
result files via <code>results &lt;- read_results(files, meta_cond)</code> where 
<code>meta_cond</code> is a vector of character strings that provides a descriptive 
name for each file in <code>files</code>.</p>
<p>Check <code><a href="../articles/JACUSA2helper.html">vignette("JACUSA2helper", "JACUSA2helper")</a></code> for a general 
introduction and <code>"JACUSA2helper", "JACUSA2helper"</code> for details about 
meta conditions.</p>
    </div>
    <div class="section level2">
    <h2 id="read-write-functions">read/write functions<a class="anchor" aria-label="anchor" href="#read-write-functions"></a></h2>
    

<p>See:</p><dl><dt>read_result</dt>
<dd><p>Reads and unpacks a JACUSA2 result file.</p></dd>

  <dt>read_results</dt>
<dd><p>Allows to combine multiple result files and distinguish them with meta conditions.</p></dd>

  <dt>write_bedGraph</dt>
<dd><p>Writes a vector of values as bedGraph file.</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="helper-functions">Helper functions<a class="anchor" aria-label="anchor" href="#helper-functions"></a></h2>
    

<p>See:</p><dl><dt>arrest_rate</dt>
<dd><p>Calculates arrest rate from base call counts (arest, through).</p></dd>

  <dt>base_count</dt>
<dd><p>Calculates the bumber of observed base calls.</p></dd>

  <dt>base_sub</dt>
<dd><p>Calculates base substitution.</p></dd>

  <dt>base_ratio</dt>
<dd><p>Calculates base call ratios.</p></dd>

  <dt>non_ref_ratio</dt>
<dd><p>Calculates non reference base ratio to JACUSA2 result object.</p></dd>

  <dt>sub_ratio</dt>
<dd><p>Calculates base substitution ratio for all bases to a JACUSA2 result object.</p></dd>



</dl></div>
    <div class="section level2">
    <h2 id="filter-functions">filter functions<a class="anchor" aria-label="anchor" href="#filter-functions"></a></h2>
    

<p>This function set enables filtering by read coverage or enforcing a minimal 
number of variant base calls per sample.</p>
<p>See:</p><dl><dt>robust</dt>
<dd><p>Retains sites that are robust in one feature. The feature can be observed in all replicates of at least one condition.</p></dd>

  <dt>filter_artefact</dt>
<dd><p>Remove sites that have been marked as an artefact.</p></dd>

 <dt>All</dt>
<dd><p>Helper function</p></dd>

 <dt>Any</dt>
<dd><p>Helper function</p></dd>

 <dt>lapply_cond</dt>
<dd><p>lapply wrapper - applies function to all conditions.</p></dd>

 <dt>lapply_repl</dt>
<dd><p>lapply wrapper - applies function to all replicates.</p></dd>

 <dt>mapply_repl</dt>
<dd><p>mapply wrapper - applies function to all replicates.</p></dd>


</dl></div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Michael Piechotta, <a href="http://dieterichlab.org/" class="external-link">Dieterich Lab</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

