<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="JACUSA2helper">
<title>Analysis of Nanopore HEK293 with JACUSA2helper • JACUSA2helper</title>
<script src="../../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../../pkgdown.js"></script><meta property="og:title" content="Analysis of Nanopore HEK293 with JACUSA2helper">
<meta property="og:description" content="JACUSA2helper">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../../index.html">JACUSA2helper</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.9.9.9600</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../../articles/JACUSA2helper.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../../articles/JACUSA2helper-meta-conditions.html">Introduction to meta conditions with JACUSA2helper</a>
    <a class="dropdown-item" href="../../articles/web_only/JACUSA2helper-nanopore.html">Analysis of Nanopore HEK293 with JACUSA2helper</a>
    <a class="dropdown-item" href="../../articles/JACUSA2helper-dart-seq.html">Analysis of DART-seq with JACUSA2helper</a>
    <a class="dropdown-item" href="../../articles/JACUSA2helper-mazf.html">Analysis of m6A mapping by MazF with JACUSA2helper</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/dieterich-lab/JACUSA2helper/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Analysis of Nanopore HEK293 with JACUSA2helper</h1>
                        <h4 data-toc-skip class="author">Christoph
Dieterich, Michael Piechotta</h4>
            
            <h4 data-toc-skip class="date">2023-01-18</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/dieterich-lab/JACUSA2helper/blob/HEAD/vignettes/web_only/JACUSA2helper-nanopore.Rmd" class="external-link"><code>vignettes/web_only/JACUSA2helper-nanopore.Rmd</code></a></small>
      <div class="d-none name"><code>JACUSA2helper-nanopore.Rmd</code></div>
    </div>

    
    
<ul>
<li>This vignette requires the latest version of JACUSA2helper - please
update*</li>
</ul>
<p>In the following, the workflow for use case 3 from <span class="citation">Piechotta et al. (<a href="#ref-Piechotta2021" role="doc-biblioref">2021</a>)</span> is presented. Data for <a href="https://zenodo.org/record/5930729#.YfmTDFvMIUG" class="external-link">use cases 1-3</a>
can be downloaded to repeat the analysis.</p>
<p>The following packages need to be loaded:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dieterich-lab/JACUSA2helper" class="external-link">JACUSA2helper</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/BSgenome" class="external-link">BSgenome</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomeInfoDb" class="external-link">GenomeInfoDb</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://expasy.org/tools/pROC/" class="external-link">pROC</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">VennDiagram</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://renozao.github.io/NMF/" class="external-link">NMF</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="non-negative-matrix-factorization">Non-negative Matrix Factorization<a class="anchor" aria-label="anchor" href="#non-negative-matrix-factorization"></a>
</h2>
<p>Here, we briefly introduce Non-negative Matrix Factorization (NMF).
This section uses the notation from <a href="https://cran.r-project.org/web/packages/NMF/vignettes/NMF-vignette.pdf" class="external-link">NMF
vignette</a>.</p>
<p>Non-negative Matrix Factorization tries to find an approximation of:
<span class="math display">\[X \approx W H,\]</span> where:</p>
<ul>
<li>
<span class="math inline">\(X\)</span> is a non-negative <span class="math inline">\(n \times p\)</span> matrix,</li>
<li>
<span class="math inline">\(W\)</span> is a non-negative <span class="math inline">\(n \times r\)</span> matrix (basis matrix),
and</li>
<li>
<span class="math inline">\(H\)</span> is a non-negative <span class="math inline">\(r \times p\)</span> matrix (coefficient
matrix).</li>
</ul>
<p><span class="math inline">\(r\)</span> is called the factorization
rank.</p>
<p>Check <a href="https://cran.r-project.org/web/packages/NMF/vignettes/NMF-vignette.pdf" class="external-link">NMF
vignette</a> for details on optimization and algorithms.</p>
<p>In the following, we will explain how to convert JACUSA2 output to an
appropriate matrix <span class="math inline">\(X\)</span> and perform
NMF.</p>
</div>
<div class="section level2">
<h2 id="general-workflow">General workflow<a class="anchor" aria-label="anchor" href="#general-workflow"></a>
</h2>
<p>Use <code>read_result</code> to read JACUSA2 output and filter the
data set.</p>
<p>To repeat the Nanopore analysis from <span class="citation">Piechotta
et al. (<a href="#ref-Piechotta2021" role="doc-biblioref">2021</a>)</span>, download the following data:</p>
<ul>
<li>
<a href="https://zenodo.org/record/5940218/files/WT_vs_KO_call2_result.out.gz?download=1" class="external-link">WT
vs. KO</a> and</li>
<li>
<a href="https://zenodo.org/record/5940218/files/WT100_vs_WT0_call2_result.out.gz?download=1" class="external-link">WT100
vs. WT0</a>.</li>
</ul>
<p>Depending on your bandwidth the Download might take some time.
Download the data, start R and change into the directory where you have
saved the files from zenodo. Make sure, you have approx. 20GB of main
memory available.</p>
<div class="section level3">
<h3 id="pre-processing">Pre-processing<a class="anchor" aria-label="anchor" href="#pre-processing"></a>
</h3>
<p>We start to organize the data:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># files to read</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"WT_vs_KO_call2_result.out.gz"</span>,</span>
<span>  <span class="st">"WT100_vs_WT0_call2_result.out.gz"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># descriptions corresponding to files </span></span>
<span><span class="va">meta_conds</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"WT_vs_KO"</span>,</span>
<span>  <span class="st">"WT100_vs_WT0"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><code>meta_conds</code> is used as a concise description of each file
to distinguish the JACUSA2 outputs.</p>
<p>The pre-processing workflow for one file can be summarized into the
following steps:</p>
<dl>
<dt>Read</dt>
<dd>
Read JACUSA2 output.
</dd>
<dt>Filter</dt>
<dd>
Employ coverage and other filters to remove not interesting sites.
</dd>
<dt>Add meta condition</dt>
<dd>
Add concise description of each file.
</dd>
<dt>Reduce</dt>
<dd>
Select relevant data column.
</dd>
</dl>
<p>We explain the pre-processing workflow for one file and generalize to
arbitrary number of files (here two).</p>
<!-- TODO mention `read_results` 
### `read_results`

An other option to read the data set is via `read_results` - see `vignette(Introduction to meta conditions with JACUSA2helper)`.
Does not apply here because the data has different number of replicates.
This approach will consume much more main memory. We recommend to read big data sets serially with `read_result` and combine them afterwards.
-->
<div class="section level4">
<h4 id="read">Read<a class="anchor" aria-label="anchor" href="#read"></a>
</h4>
<p>The underlying function of <code>read_result</code> is
<code><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">data.table::fread</a></code> - check the respective help page for
details on additional options. Depending on your machine increase
<code>nThreads</code> to use more threads to parse a file.</p>
<p>The “info” field contains meta information for sites, such as
detailed INDEL statistics. To save memory, we manually unpack the “info”
field and select the following keys:</p>
<ul>
<li>
<code>insertion_score</code> and</li>
<li>
<code>deletion_score</code>.</li>
</ul>
<p>We will use these score values during model training.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">i</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="co"># WT_vs_KO</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">files</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/home/michael/TODO/JACUSA2helper/WT_vs_KO_call2_result.out.gz"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">meta_conds</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "WT_vs_KO"</span></span>
<span></span>
<span><span class="va">wt_vs_ko_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/read_result.html">read_result</a></span><span class="op">(</span><span class="va">files</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, nThread <span class="op">=</span> <span class="fl">2</span>, unpack <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"insertion_score"</span>, <span class="st">"deletion_score"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="filter">Filter<a class="anchor" aria-label="anchor" href="#filter"></a>
</h4>
<p>The JACUSA2 output files from Zenodo have been already filtered to
retain sites with coverage <span class="math inline">\(&gt; 4\)</span>
in all BAM files.</p>
<p>We use the following filter:</p>
<ul>
<li>retain sites on chromosome 1-22, MT, and X.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Sites BEFORE filtering: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">wt_vs_ko_res</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Sites BEFORE filtering: 8891484"</span></span>
<span></span>
<span><span class="co"># reduce set of known sequences</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">wt_vs_ko_res</span>, pruning.mode <span class="op">=</span> <span class="st">"coarse"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">"MT"</span>, <span class="st">"X"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">wt_vs_ko_filtered</span> <span class="op">&lt;-</span> <span class="va">wt_vs_ko_res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>    <span class="va">seqnames</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span>, <span class="st">"MT"</span>, <span class="st">"X"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Sites AFTER filtering: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">wt_vs_ko_filtered</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Sites AFTER filtering: 8880996"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="add-meta-condition">Add meta condition<a class="anchor" aria-label="anchor" href="#add-meta-condition"></a>
</h4>
<p>We add a concise description of files to each result object.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">wt_vs_ko_filtered</span><span class="op">)</span><span class="op">$</span><span class="va">meta_cond</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">meta_conds</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">meta_conds</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">wt_vs_ko_filtered</span><span class="op">$</span><span class="va">meta_cond</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     WT_vs_KO WT100_vs_WT0 </span></span>
<span><span class="co">#&gt;      8880996            0</span></span></code></pre></div>
<p><code>WT100_vs_WT0</code> is zero because we haven’t added the
respective file yet.</p>
<div class="section level5">
<h5 id="add-reference-sequence-information">Add reference sequence information<a class="anchor" aria-label="anchor" href="#add-reference-sequence-information"></a>
</h5>
<p>Sequence lengths are necessary to check if coordinates are valid.
Initially, a JACUSA2 result object has no information about the
underlying reference sequence.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">wt_vs_ko_res</span><span class="op">)</span></span>
<span><span class="co">#&gt;  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 MT  X </span></span>
<span><span class="co">#&gt; NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA</span></span></code></pre></div>
<p>We can either specify the genome (see
<code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">?GenomeInfoDb::seqinfo</a></code>) or add reference sequence
information with <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths()</a></code>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load the package that corresponds to your genome</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.NCBI.GRCh38"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">wt_vs_ko_res</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.NCBI.GRCh38</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">wt_vs_ko_res</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">wt_vs_ko_res</span><span class="op">)</span></span>
<span><span class="co">#&gt;         1         2         3         4         5         6         7         8 </span></span>
<span><span class="co">#&gt; 248956422 242193529 198295559 190214555 181538259 170805979 159345973 145138636 </span></span>
<span><span class="co">#&gt;         9        10        11        12        13        14        15        16 </span></span>
<span><span class="co">#&gt; 138394717 133797422 135086622 133275309 114364328 107043718 101991189  90338345 </span></span>
<span><span class="co">#&gt;        17        18        19        20        21        22        MT         X </span></span>
<span><span class="co">#&gt;  83257441  80373285  58617616  64444167  46709983  50818468     16569 156040895</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="create-matrix-x">Create matrix <span class="math inline">\(X\)</span><a class="anchor" aria-label="anchor" href="#create-matrix-x"></a>
</h4>
<p>We continue to reduce the data set by selecting only relevant columns
and renaming the “score” column to “call2_score” for consistency: Given
a JACUSA2 result object, the function <code><a href="../../reference/create_data.html">create_data()</a></code> will
create a data matrix, that will be used in NMF. In brief, relevant score
columns are rearranged and a region or context around each Adenin site
is created.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_matrix</span> <span class="op">=</span> <span class="fu"><a href="../../reference/create_data.html">create_data</a></span><span class="op">(</span><span class="va">wt_vs_ko_filtered</span><span class="op">)</span></span>
<span><span class="va">data_matrix</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">data_matrix</span><span class="op">[</span><span class="va">data_matrix</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "WT_vs_KO_call2_score_1"     "WT_vs_KO_call2_score_2"    </span></span>
<span><span class="co">#&gt;  [3] "WT_vs_KO_call2_score_3"     "WT_vs_KO_call2_score_4"    </span></span>
<span><span class="co">#&gt;  [5] "WT_vs_KO_call2_score_5"     "WT_vs_KO_deletion_score_1" </span></span>
<span><span class="co">#&gt;  [7] "WT_vs_KO_deletion_score_2"  "WT_vs_KO_deletion_score_3" </span></span>
<span><span class="co">#&gt;  [9] "WT_vs_KO_deletion_score_4"  "WT_vs_KO_deletion_score_5" </span></span>
<span><span class="co">#&gt; [11] "WT_vs_KO_insertion_score_1" "WT_vs_KO_insertion_score_2"</span></span>
<span><span class="co">#&gt; [13] "WT_vs_KO_insertion_score_3" "WT_vs_KO_insertion_score_4"</span></span>
<span><span class="co">#&gt; [15] "WT_vs_KO_insertion_score_5"</span></span></code></pre></div>
<p>This operation concludes the pre-processing workflow for one
file.</p>
</div>
<div class="section level4">
<h4 id="generalize">Generalize<a class="anchor" aria-label="anchor" href="#generalize"></a>
</h4>
<p>So far, we have shown a detailed description of pre-processing one
file. In the following, we provide code for arbitrary number of
results/experiments/meta conditions:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># pre-processing workflow for arbitrary number of files and meta conditions</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">file</span>, <span class="va">meta_cond</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/read_result.html">read_result</a></span><span class="op">(</span><span class="va">file</span>, nThread <span class="op">=</span> <span class="fl">3</span>, unpack <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"insertion_score"</span>, <span class="st">"deletion_score"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">seqnames</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span>, <span class="st">"MT"</span>, <span class="st">"X"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">result</span>, pruning.mode <span class="op">=</span> <span class="st">"coarse"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">"MT"</span>, <span class="st">"X"</span><span class="op">)</span></span>
<span>    <span class="va">result</span> <span class="op">&lt;-</span> <span class="va">result</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"score"</span>, <span class="st">"insertion_score"</span>, <span class="st">"deletion_score"</span>, <span class="st">"filter"</span>, <span class="st">"ref"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span>    <span class="co"># add file specific condition</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span><span class="op">$</span><span class="va">meta_cond</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">meta_cond</span>, <span class="va">meta_conds</span><span class="op">)</span></span>
<span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span>
<span>  <span class="op">}</span>, <span class="va">files</span>, <span class="va">meta_conds</span>, SIMPLIFY <span class="op">=</span> <span class="cn">FALSE</span>, USE.NAMES <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We need to convert <code>results</code> to a GRanges object and we
add sequence information.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># convert and concatenate GenomicRanges</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add reference sequence information</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.NCBI.GRCh38</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># retained sites per file / meta condition</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">meta_cond</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     WT_vs_KO WT100_vs_WT0 </span></span>
<span><span class="co">#&gt;      8880996     11147905</span></span></code></pre></div>
<p><code>results</code> now consists of filtered JACUSA2 calls for sites
from all <code>files</code> and <code>meta_conds</code>.</p>
<p>The overlap of called sites from each experiment can be investigated
with the following:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add unique ID for a site: contig:start-end:strand</span></span>
<span><span class="va">results</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># venn diagram of sites that are shared between files</span></span>
<span><span class="va">meta_cond_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/VennDiagram/man/venn.diagram.html" class="external-link">venn.diagram</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">id</span>, <span class="va">results</span><span class="op">$</span><span class="va">meta_cond</span>, <span class="va">c</span><span class="op">)</span>,</span>
<span>  filename <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  lwd <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  cex <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  fontfamily <span class="op">=</span> <span class="st">"sans"</span>,</span>
<span>  cat.cex <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  cat.default.pos <span class="op">=</span> <span class="st">"outer"</span>,</span>
<span>  cat.fontfamily <span class="op">=</span> <span class="st">"sans"</span>,</span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/grid.newpage.html" class="external-link">grid.newpage</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/grid.draw.html" class="external-link">grid.draw</a></span><span class="op">(</span><span class="va">meta_cond_plt</span><span class="op">)</span></span></code></pre></div>
<p><img src="JACUSA2helper-nanopore_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="create-data-matrix">Create data matrix<a class="anchor" aria-label="anchor" href="#create-data-matrix"></a>
</h3>
<p>We use <code><a href="../../reference/create_data.html">create_data()</a></code> to create the data matrix that will
be used to learn the model. We add a context of 2nt around each site
creating a region (-NN<strong>A</strong>NN-) where the putative
modification site is positioned in the middle (= position 3) of the
motif. Internally, <code><a href="../../reference/create_data.html">create_data()</a></code> removes sites that are
within homopolymers (JACUSA2 filter flag: “Y”). Each column name of the
data matrix has the following format:
<code>{meta_cond}_{score}_{position}</code>, where:</p>
<dl>
<dt><code>meta_cond</code></dt>
<dd>
is meta condition information (<span class="math inline">\(meta_cond \in
{WT_vs_KO, WT100_vs_WT}\)</span>) <code>score</code>
</dd>
<dd>
type of score <span class="math inline">\(score \in {call2, insertion,
or deletion}\)</span> <code>position</code>
</dd>
<dd>
position information within the region (<span class="math inline">\(position \in {1, ..., 5}\)</span>).
</dd>
</dl>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/create_data.html">create_data</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set sensible defaults</span></span>
<span><span class="va">data_matrix</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">data_matrix</span><span class="op">[</span><span class="va">data_matrix</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span></code></pre></div>
<p>Next, we filter the data matrix and require that each region has to
be covered in both experiments.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># contains score sum for each meta condition</span></span>
<span><span class="va">score_sums</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rowsum.html" class="external-link">rowsum</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/strcapture.html" class="external-link">strcapture</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"^("</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">meta_conds</span>, collapse <span class="op">=</span> <span class="st">"|"</span><span class="op">)</span>, <span class="st">")"</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>meta_cond <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">$</span><span class="va">meta_cond</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># score sum must be &gt; 0 in all experiments</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">score_sums</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">meta_conds</span><span class="op">)</span></span>
<span><span class="va">data_matrix</span> <span class="op">&lt;-</span> <span class="va">data_matrix</span><span class="op">[</span><span class="va">keep</span>, <span class="op">]</span></span></code></pre></div>
<p>We will continue adding meta information to the data matrix to study
the properties of sites.</p>
<div class="section level4">
<h4 id="add-sequence">Add sequence<a class="anchor" aria-label="anchor" href="#add-sequence"></a>
</h4>
<p>In order to add information if a site is contained in a DRACH motif,
we need to retrieve the sequence context for a site.</p>
<p>If you have a custom FASTA sequence, use
<code><a href="https://rdrr.io/pkg/Rsamtools/man/FaFile-class.html" class="external-link">Rsamtools::FaFile</a></code> to load the FASTA file. Otherwise, load a
genome from via <a href="https://bioconductor.org/packages/release/bioc/html/BSgenome.html" class="external-link">BSgenome</a>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># retrieve sequence and convert to character vector</span></span>
<span><span class="va">data_matrix</span><span class="op">$</span><span class="va">motif</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/getSeq.html" class="external-link">getSeq</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.NCBI.GRCh38</span>, <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># number of top 10 motifs</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">$</span><span class="va">motif</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; CCAGC GCAGC GGAGG CCAGG CCACC AGAAG GCAGG TGAAG GGAAG CCAAG </span></span>
<span><span class="co">#&gt; 28807 28624 23477 23187 22785 22400 21197 20452 19651 19590</span></span></code></pre></div>
<p>Next, we add an indicator variable if the DRACH motif
([AGT][AG]AC[ACT]) is present:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_matrix</span><span class="op">$</span><span class="va">DRACH</span> <span class="op">&lt;-</span> <span class="st">"no"</span></span>
<span><span class="va">data_matrix</span><span class="op">$</span><span class="va">DRACH</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"[AGT][AG]AC[ACT]"</span>, <span class="va">data_matrix</span><span class="op">$</span><span class="va">motif</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"yes"</span></span>
<span></span>
<span><span class="va">tbl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">$</span><span class="va">DRACH</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">tbl</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">tbl</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>    <span class="st">" ("</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>      <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/label_number.html" class="external-link">label_comma</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">tbl</span><span class="op">)</span><span class="op">)</span>, </span>
<span>      <span class="st">"; "</span>, </span>
<span>      <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent_format</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">tbl</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">tbl</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span>, </span>
<span>    <span class="st">")"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># plot distribution of DRACH motifs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/pie.html" class="external-link">pie</a></span><span class="op">(</span><span class="va">tbl</span>, mai <span class="op">=</span> <span class="st">"DRACH motif present"</span><span class="op">)</span></span></code></pre></div>
<p><img src="JACUSA2helper-nanopore_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p>Use <code>saveRDS(data_matrix, file = "data_matrix.rds")</code> to
store the current state. Use
<code>data_matrix &lt;- readRDS("data_matrix.rds")</code> to restore
previous state.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="use-nmf-on-nanopore-data">Use NMF on Nanopore data<a class="anchor" aria-label="anchor" href="#use-nmf-on-nanopore-data"></a>
</h2>
<p>The learning part can be omitted. Use
<code>data(m6a_nmf_results)</code> from <span class="citation">Piechotta
et al. (<a href="#ref-Piechotta2021" role="doc-biblioref">2021</a>)</span> and and continue to <a href="#visualize-nmf">Visualize NMF</a>.</p>
<p>Continue reading <a href="#overlap-with-miclip-data">Overlap with
miCLIP data</a>, to understand how object <code>m6a_nmf_results</code>
can be created.</p>
<div class="section level3">
<h3 id="overlap-with-miclip-data">Overlap with miCLIP data<a class="anchor" aria-label="anchor" href="#overlap-with-miclip-data"></a>
</h3>
<p>Next, we use existing miCLIP data to train the model.</p>
<p>The data can be investigated with: <code>data(m6a_miclip)</code>. It
consists of 3 experiments:</p>
<ul>
<li>
<span class="citation">Körtel et al. (<a href="#ref-kortel_deep_2021" role="doc-biblioref">2021</a>)</span>,</li>
<li>
<span class="citation">Koh, Goh, and Goh (<a href="#ref-koh_atlas_2019" role="doc-biblioref">2019</a>)</span>,
and</li>
<li>
<span class="citation">Boulias et al. (<a href="#ref-boulias_identification_2019" role="doc-biblioref">2019</a>)</span> .</li>
</ul>
<p>The column “experiments” indicates the source of a site.</p>
<p>Plot venn diagram of shared CLIP sites:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">miclip_plt</span> <span class="op">&lt;-</span> <span class="va">m6a_miclip</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>site <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate_rows.html" class="external-link">separate_rows</a></span><span class="op">(</span><span class="va">experiments</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="va">site</span>, <span class="va">experiments</span>, <span class="va">list</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/VennDiagram/man/venn.diagram.html" class="external-link">venn.diagram</a></span><span class="op">(</span></span>
<span>    filename <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    cex <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>    fontfamily <span class="op">=</span> <span class="st">"sans"</span>,</span>
<span>    cat.cex <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>    cat.default.pos <span class="op">=</span> <span class="st">"outer"</span>,</span>
<span>    cat.fontfamily <span class="op">=</span> <span class="st">"sans"</span>,</span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/grid.newpage.html" class="external-link">grid.newpage</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/grid.draw.html" class="external-link">grid.draw</a></span><span class="op">(</span><span class="va">miclip_plt</span><span class="op">)</span></span></code></pre></div>
<p><img src="JACUSA2helper-nanopore_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>We add miCLIP information to <code>data_matrix</code>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">m6a_miclip</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add region around site to miCLIP data</span></span>
<span><span class="va">m6a_miclip_region</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/extend.html">extend</a></span><span class="op">(</span><span class="va">m6a_miclip</span>, left <span class="op">=</span> <span class="fl">2</span>, right <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">m6a_miclip_region</span><span class="op">$</span><span class="va">experiments</span> <span class="op">&lt;-</span> <span class="va">m6a_miclip</span><span class="op">$</span><span class="va">experiments</span></span>
<span></span>
<span><span class="va">miclip_covered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">subsetByOverlaps</a></span><span class="op">(</span></span>
<span>  <span class="va">m6a_miclip_region</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"equal"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">data_matrix</span><span class="op">$</span><span class="va">experiments</span> <span class="op">&lt;-</span> <span class="st">"-"</span></span>
<span><span class="va">data_matrix</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">miclip_covered</span><span class="op">)</span>, <span class="st">"experiments"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">miclip_covered</span><span class="op">$</span><span class="va">experiments</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">$</span><span class="va">experiments</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                   -             Boulias     Boulias,Koertel Boulias,Koertel,Koh </span></span>
<span><span class="co">#&gt;             2255751                8490                4966                2916 </span></span>
<span><span class="co">#&gt;         Boulias,Koh             Koertel         Koertel,Koh                 Koh </span></span>
<span><span class="co">#&gt;                 850                7831                2249                2968</span></span></code></pre></div>
<p>Next, we create the data matrix (corresponds to <span class="math inline">\(X\)</span> in <span class="math inline">\(N
\approx W H\)</span>) of sites that are contained in all three data
sets: Boulias, Koertel, and Koh. It will be used to learn the
modification patterns:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># retain sites that overlap with 3 miCLIP experiments</span></span>
<span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="va">data_matrix</span><span class="op">$</span><span class="va">experiments</span> <span class="op">==</span> <span class="st">"Boulias,Koertel,Koh"</span></span>
<span><span class="va">train_matrix</span> <span class="op">&lt;-</span> <span class="va">data_matrix</span><span class="op">[</span><span class="va">keep</span>, <span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">train_matrix</span><span class="op">$</span><span class="va">DRACH</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  yes </span></span>
<span><span class="co">#&gt; 2916</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="compute-factorizaion-ranks">Compute factorizaion rank(s)<a class="anchor" aria-label="anchor" href="#compute-factorizaion-ranks"></a>
</h3>
<p>An important parameter in NMF is the factorization rank <span class="math inline">\(r\)</span> that defines the number of features to
approximate <span class="math inline">\(X\)</span> - a similar parameter
such as the number of clusters in the k-means algorithm. We will compute
multiple factorizations with different values for <span class="math inline">\(r\)</span> and use surrogate measures to find an
appropriate value.</p>
<p>Make sure that the data/train matrix (<span class="math inline">\(X\)</span>) contains only numeric values before
you start the factorization of <span class="math inline">\(X\)</span> to
<span class="math inline">\(W H\)</span>. The functions that does the
calculations in JACUSA2helper is <code>nmf_learn(x, mods)</code>, where
<code>x</code> is the data/train matrix and mods is a GRanges object of
known modifications.</p>
<p>Internally, JACUSA2helper uses the NMF package. Check <a href="https://cran.r-project.org/web/packages/NMF/vignettes/NMF-vignette.pdf" class="external-link">NMF
vignette</a> for details on parameters. The default parameters in
JACUSA2helper are set to
<code>nmf_args = list(rank = 2:10, nrun = 10, seed = 123456, .opt = "vp4")</code>.
For example argument <code>.opt=pv3</code> instructs to use 3 cores -
adjust the number according to your machine.</p>
<p>When using the default parameters, factorizations for rank <span class="math inline">\(r \in {2, ..., 10}\)</span> will be computed:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>nmf_results <span class="ot">&lt;-</span> <span class="fu">learn_nmf</span>(</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  train_matrix <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(motif, DRACH, experiments))</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 2  ... + measures ... OK</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 3  ... + measures ... OK</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 4  ... + measures ... OK</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 5  ... + measures ... OK</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 6  ... + measures ... OK</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 7  ... + measures ... OK</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 8  ... + measures ... OK</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 9  ... + measures ... OK</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 10  ... + measures ... OK</span></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>Runs<span class="sc">:</span> <span class="er">|</span>                                                        </span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>Runs<span class="sc">:</span> <span class="er">|</span>                                                  <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>Runs<span class="sc">:</span> <span class="er">|</span>                                                        </span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>Runs<span class="sc">:</span> <span class="er">|==================================================|</span> <span class="dv">100</span>%</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; System time:</span></span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    user  system elapsed </span></span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  27.409   3.813  22.118</span></span></code></pre></div>
<p>The result <code>nmf_results</code> is a list that consists of the
following data:</p>
<dl>
<dt><code>estim_r</code></dt>
<dd>
Factorizations for <code>x</code> <code>nmf_matrix</code>
</dd>
<dd>
Factorization of <code>x</code> with <code>r</code> that maximizes
<code>estim_r$measures$silhouette.consensus</code> and
<code>estim_r$measures$cophenetic</code> <code>chosen_rank</code>
</dd>
<dd>
Optimal rank <code>chosen_pattern</code>
</dd>
<dd>
The pattern that has the highest score
</dd>
</dl>
<p>In the next steps, we will evaluate the learned model by comparing it
against a null model.</p>
<div class="section level4">
<h4 id="visualize-estimation-of-the-ranks">Visualize estimation of the ranks<a class="anchor" aria-label="anchor" href="#visualize-estimation-of-the-ranks"></a>
</h4>
<p>We permutate the initial data and learn a randomized model:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>nmf_random <span class="ot">&lt;-</span> <span class="fu">learn_nmf</span>(</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  train_matrix <span class="sc">%&gt;%</span> </span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(motif, DRACH, experiments)) <span class="sc">%&gt;%</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">randomize</span>()</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 2  ... + measures ... OK</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 3  ... + measures ... OK</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 4  ... + measures ... OK</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 5  ... + measures ... OK</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 6  ... + measures ... OK</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 7  ... + measures ... OK</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 8  ... + measures ... OK</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 9  ... + measures ... OK</span></span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Compute NMF rank= 10  ... + measures ... OK</span></span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>Runs<span class="sc">:</span> <span class="er">|</span>                                                        </span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>Runs<span class="sc">:</span> <span class="er">|</span>                                                  <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>Runs<span class="sc">:</span> <span class="er">|</span>                                                        </span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>Runs<span class="sc">:</span> <span class="er">|==================================================|</span> <span class="dv">100</span>%</span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; System time:</span></span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    user  system elapsed </span></span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  25.535   3.912  20.581</span></span></code></pre></div>
<p>In the following, we will compare the results for different
factorizations on the original and randomized data:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># compare factorization on original and randomized data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">nmf_results</span><span class="op">$</span><span class="va">estim_r</span>, <span class="va">nmf_random</span><span class="op">$</span><span class="va">estim_r</span><span class="op">)</span></span></code></pre></div>
<p><img src="JACUSA2helper-nanopore_files/figure-html/unnamed-chunk-24-1.png" width="700">
The optimal value of <span class="math inline">\(r\)</span> is the
minimum rank for which rank <code>Silhouette</code> and
<code>Cophenetic</code> is maximized.</p>
</div>
</div>
<div class="section level3">
<h3 id="visualize-nmf">Visualize NMF<a class="anchor" aria-label="anchor" href="#visualize-nmf"></a>
</h3>
<div class="section level4">
<h4 id="basis-components">Basis components<a class="anchor" aria-label="anchor" href="#basis-components"></a>
</h4>
<p>Plot properties of <span class="math inline">\(W\)</span>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/NMF/man/heatmaps.html" class="external-link">basismap</a></span><span class="op">(</span><span class="va">nmf_results</span><span class="op">$</span><span class="va">nmf_matrix</span><span class="op">)</span></span></code></pre></div>
<p><img src="JACUSA2helper-nanopore_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="mixture-components">Mixture components<a class="anchor" aria-label="anchor" href="#mixture-components"></a>
</h4>
<p>Plot properties of <span class="math inline">\(H\)</span>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/NMF/man/heatmaps.html" class="external-link">coefmap</a></span><span class="op">(</span><span class="va">nmf_results</span><span class="op">$</span><span class="va">nmf_matrix</span><span class="op">)</span></span></code></pre></div>
<p><img src="JACUSA2helper-nanopore_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="visualize-patterns">Visualize patterns<a class="anchor" aria-label="anchor" href="#visualize-patterns"></a>
</h3>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">w</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NMF/man/basis-coef-methods.html" class="external-link">basis</a></span><span class="op">(</span><span class="va">nmf_results</span><span class="op">$</span><span class="va">nmf_matrix</span><span class="op">)</span></span>
<span><span class="co"># Table of number of best hits for each pattern"</span></span>
<span><span class="va">pattern_instances</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">w</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">x</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span><span class="va">pattern_instance_tbl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">pattern_instances</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span></span>
<span>  height <span class="op">=</span> <span class="va">pattern_instance_tbl</span>,</span>
<span>  names <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">w</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"NMF Patterns Scoring"</span>, </span>
<span>  xlab <span class="op">=</span> <span class="st">"Patterns"</span>, </span>
<span>  ylab <span class="op">=</span> <span class="st">"Membership Score (basis matrix)"</span>,</span>
<span>  ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/pretty.html" class="external-link">pretty</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">pattern_instance_tbl</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="JACUSA2helper-nanopore_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<p>From this plot, we see that the highest score can observed for
pattern:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nmf_results</span><span class="op">$</span><span class="va">chosen_pattern</span></span>
<span><span class="co">#&gt; [1] 5</span></span></code></pre></div>
<p>Next, we visualize the sequence logos that correspond to calculated
patterns in NMF.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/coef.html" class="external-link">coef</a></span><span class="op">(</span><span class="va">nmf_results</span><span class="op">$</span><span class="va">nmf_matrix</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">h</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">k</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">region_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">train_matrix</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">pattern_instances</span> <span class="op">==</span> <span class="va">k</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">train_matrix</span><span class="op">[</span><span class="va">region_index</span>, <span class="st">"motif"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sequences</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Pattern "</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">h</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">ggseqlogo</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggseqlogo/man/ggseqlogo.html" class="external-link">ggseqlogo</a></span><span class="op">(</span><span class="va">sequences</span><span class="op">)</span></span></code></pre></div>
<p><img src="JACUSA2helper-nanopore_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
<p>Finally, we plot profiles for each pattern and experiment (meta
condition):</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">h</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span></span>
<span>    <span class="op">-</span><span class="va">pattern</span>, </span>
<span>    names_to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Exp"</span>, <span class="st">"Score"</span>, <span class="st">"Pos"</span><span class="op">)</span>, </span>
<span>    names_pattern <span class="op">=</span> <span class="st">"(WT_vs_KO|WT100_vs_WT0)_(call2_score|deletion_score|insertion_score)_([1-5]+)"</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>score <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_score$"</span>, <span class="st">""</span>, <span class="va">Score</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Pos</span>, y <span class="op">=</span> <span class="va">value</span>, fill <span class="op">=</span> <span class="va">Score</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Position in motif\n-NN A NN-"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">Exp</span> <span class="op">~</span> <span class="va">pattern</span>, labeller <span class="op">=</span> <span class="va">label_both</span><span class="op">)</span></span></code></pre></div>
<p><img src="JACUSA2helper-nanopore_files/figure-html/unnamed-chunk-30-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="prediction">Prediction<a class="anchor" aria-label="anchor" href="#prediction"></a>
</h3>
<p>We evaluate the factorization on the previously introduced miCLIP
data. We use all sites. WARNING! Make sure that the order of columns is
correct - use <code><a href="../../reference/create_data.html">create_data()</a></code>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">score_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/predict_mods.html">predict_mods</a></span><span class="op">(</span></span>
<span>  <span class="va">data_matrix</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">motif</span>, <span class="va">DRACH</span>, <span class="va">experiments</span><span class="op">)</span><span class="op">)</span>, </span>
<span>  <span class="va">nmf_results</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><code>score_matrix</code> consists of columns scores for each pattern
and the <code>score</code> for the chosen pattern (in this case:
<code>NMF5</code> = <code>score</code>). We add meta information from
<code>data_matrix</code> to <code>score_matrix</code>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">score_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">score_matrix</span>, <span class="va">data_matrix</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"motif"</span>, <span class="st">"DRACH"</span>, <span class="st">"experiments"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">score_matrix</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "NMF1"        "NMF2"        "NMF3"        "NMF4"        "NMF5"       </span></span>
<span><span class="co">#&gt; [6] "score"       "motif"       "DRACH"       "experiments"</span></span></code></pre></div>
<p>Sort <code>score_matrix</code> by column <code>score</code> - the
best predictions will appear in the upper part of matrix.</p>
</div>
<div class="section level3">
<h3 id="evaluation">Evaluation<a class="anchor" aria-label="anchor" href="#evaluation"></a>
</h3>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">score_matrix</span><span class="op">$</span><span class="va">clip</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">score_matrix</span><span class="op">$</span><span class="va">experiments</span> <span class="op">!=</span> <span class="st">"-"</span>, <span class="st">"CLIP"</span>, <span class="st">"No Clip"</span><span class="op">)</span></span>
<span><span class="va">score_matrix</span><span class="op">$</span><span class="va">type</span> <span class="op">&lt;-</span> <span class="st">"-"</span></span>
<span></span>
<span><span class="va">train</span> <span class="op">&lt;-</span> <span class="va">score_matrix</span><span class="op">$</span><span class="va">experiments</span> <span class="op">==</span> <span class="st">"Boulias,Koertel,Koh"</span></span>
<span><span class="va">score_matrix</span><span class="op">$</span><span class="va">type</span><span class="op">[</span><span class="va">train</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"train"</span></span>
<span><span class="va">score_matrix</span><span class="op">$</span><span class="va">type</span><span class="op">[</span><span class="op">!</span><span class="va">train</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"evaluate"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">score_matrix</span><span class="op">$</span><span class="va">clip</span>, <span class="va">score_matrix</span><span class="op">$</span><span class="va">type</span><span class="op">)</span></span>
<span><span class="co">#&gt;          </span></span>
<span><span class="co">#&gt;           evaluate   train</span></span>
<span><span class="co">#&gt;   CLIP       27354    2916</span></span>
<span><span class="co">#&gt;   No Clip  2255751       0</span></span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_matrix</span> <span class="op">&lt;-</span> <span class="va">score_matrix</span><span class="op">[</span><span class="va">score_matrix</span><span class="op">$</span><span class="va">type</span> <span class="op">==</span> <span class="st">"evaluate"</span>, <span class="op">]</span></span>
<span><span class="va">r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/pROC/man/roc.html" class="external-link">roc</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">test_matrix</span><span class="op">[</span>, <span class="st">"clip"</span><span class="op">]</span> <span class="op">==</span> <span class="st">"CLIP"</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>, </span>
<span>  <span class="va">test_matrix</span><span class="op">[</span>, <span class="st">"score"</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Setting levels: control = 0, case = 1</span></span>
<span><span class="co">#&gt; Setting direction: controls &lt; cases</span></span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coordinates</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/pROC/man/coords.html" class="external-link">coords</a></span><span class="op">(</span><span class="va">r</span>, x <span class="op">=</span> <span class="st">"all"</span>, input <span class="op">=</span> <span class="st">"threshold"</span>, ret <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"threshold"</span>,<span class="st">"ppv"</span>,<span class="st">"tpr"</span>,<span class="st">"tp"</span>, <span class="st">"tn"</span>,<span class="st">"fp"</span>,<span class="st">"fn"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">coordinates</span><span class="op">$</span><span class="va">sum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">coordinates</span><span class="op">$</span><span class="va">tp</span> <span class="op">+</span> <span class="va">coordinates</span><span class="op">$</span><span class="va">fp</span><span class="op">)</span></span>
<span><span class="va">coordinates</span><span class="op">$</span><span class="va">ppv</span> <span class="op">&lt;-</span> <span class="va">coordinates</span><span class="op">$</span><span class="va">ppv</span> <span class="op">*</span> <span class="fl">20</span></span>
<span><span class="va">coordinates</span> <span class="op">&lt;-</span> <span class="va">coordinates</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">coordinates</span><span class="op">$</span><span class="va">sum</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">coordinates</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">threshold</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">sum</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"#009E73"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">ppv</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"#D55E00"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span></span>
<span>    <span class="co"># Features of the first axis</span></span>
<span>    name <span class="op">=</span> <span class="st">"Log2(#predicted modified)"</span>,</span>
<span>    <span class="co"># Add a second axis and specify its features</span></span>
<span>    sec.axis <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/sec_axis.html" class="external-link">sec_axis</a></span><span class="op">(</span><span class="op">~</span><span class="va">.</span><span class="op">*</span><span class="fl">5</span>, name <span class="op">=</span> <span class="st">"PPV (miCLIP)"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span>  </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"JACUSA2 NMF Score"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>,      </span>
<span>    axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"#009E73"</span>, size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    axis.title.y.right <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"#D55E00"</span>, size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.9</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="JACUSA2helper-nanopore_files/figure-html/unnamed-chunk-36-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="prediction-on-1-experiment">Prediction on 1 experiment<a class="anchor" aria-label="anchor" href="#prediction-on-1-experiment"></a>
</h2>
<p>In the following we will show how to predict modification sites on
only one experiment.</p>
<div class="section level3">
<h3 id="prepare-data">Prepare data<a class="anchor" aria-label="anchor" href="#prepare-data"></a>
</h3>
<p>We will predict modification sites only on WT vs KO from the previous
example. If you have your own data, follow the steps above until
<code><a href="../../reference/create_data.html">create_data()</a></code>. Here, we will modify the existing
<code>data_matrix</code>:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">WT_vs_KO_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span></span>
<span>  <span class="va">data_matrix</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"^WT_vs_KO"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>,</span>
<span>  <span class="va">data_matrix</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"motif"</span>, <span class="st">"DRACH"</span>, <span class="st">"experiments"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="va">WT_vs_KO_data</span><span class="op">$</span><span class="va">clip</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">WT_vs_KO_data</span><span class="op">$</span><span class="va">experiments</span> <span class="op">!=</span> <span class="st">"-"</span>, <span class="st">"CLIP"</span>, <span class="st">"No Clip"</span><span class="op">)</span></span>
<span><span class="va">WT_vs_KO_data</span><span class="op">$</span><span class="va">type</span> <span class="op">&lt;-</span> <span class="st">"-"</span></span>
<span></span>
<span><span class="va">train</span> <span class="op">&lt;-</span> <span class="va">WT_vs_KO_data</span><span class="op">$</span><span class="va">experiments</span> <span class="op">==</span> <span class="st">"Boulias,Koertel,Koh"</span></span>
<span><span class="va">WT_vs_KO_data</span><span class="op">$</span><span class="va">type</span><span class="op">[</span><span class="va">train</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"train"</span></span>
<span><span class="va">WT_vs_KO_data</span><span class="op">$</span><span class="va">type</span><span class="op">[</span><span class="op">!</span><span class="va">train</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"evaluate"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="reduce-coef">Reduce coef<a class="anchor" aria-label="anchor" href="#reduce-coef"></a>
</h3>
<p>We need to adjust the coefficient matrix and combine the coefficients
from WT vs KO and WT100 vs WT0.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nmf_reduced</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/reduce_coef.html">reduce_coef</a></span><span class="op">(</span><span class="va">nmf_results</span>, <span class="va">meta_conds</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">nmf_reduced</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "estim_r"        "chosen_rank"    "chosen_pattern" "nmf_matrix"    </span></span>
<span><span class="co">#&gt; [5] "reduced_coef"</span></span></code></pre></div>
<p>This will calculate the mean of the coefficients and add the result
as <code>reduced_coef</code> to <code>nmf_results</code>.</p>
</div>
<div class="section level3">
<h3 id="prediction-1">Prediction<a class="anchor" aria-label="anchor" href="#prediction-1"></a>
</h3>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">WT_vs_KO_pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/predict_mods.html">predict_mods</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">WT_vs_KO_data</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">motif</span>, <span class="va">DRACH</span>, <span class="va">experiments</span>, <span class="va">clip</span>, <span class="va">type</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="va">nmf_reduced</span></span>
<span><span class="op">)</span></span>
<span><span class="va">WT_vs_KO_pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">WT_vs_KO_pred</span>, <span class="va">WT_vs_KO_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clip"</span>, <span class="st">"type"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Column <code>score</code>, can be used to sort the prediction.</p>
</div>
<div class="section level3">
<h3 id="evaluate">Evaluate<a class="anchor" aria-label="anchor" href="#evaluate"></a>
</h3>
<p>We repeat the evaluation now on WT vs KO and the reduced coefficient
matrix on.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">WT_vs_KO_test</span> <span class="op">&lt;-</span> <span class="va">WT_vs_KO_pred</span><span class="op">[</span><span class="va">WT_vs_KO_pred</span><span class="op">$</span><span class="va">type</span> <span class="op">==</span> <span class="st">"evaluate"</span>, <span class="op">]</span></span>
<span><span class="va">WT_vs_KO_roc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/pROC/man/roc.html" class="external-link">roc</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">WT_vs_KO_test</span><span class="op">[</span>, <span class="st">"clip"</span><span class="op">]</span> <span class="op">==</span> <span class="st">"CLIP"</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>, </span>
<span>  <span class="va">WT_vs_KO_test</span><span class="op">[</span>, <span class="st">"score"</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Setting levels: control = 0, case = 1</span></span>
<span><span class="co">#&gt; Setting direction: controls &lt; cases</span></span></code></pre></div>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">WT_vs_KO_coords</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/pROC/man/coords.html" class="external-link">coords</a></span><span class="op">(</span><span class="va">WT_vs_KO_roc</span>, x <span class="op">=</span> <span class="st">"all"</span>, input <span class="op">=</span> <span class="st">"threshold"</span>, ret <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"threshold"</span>,<span class="st">"ppv"</span>,<span class="st">"tpr"</span>,<span class="st">"tp"</span>, <span class="st">"tn"</span>,<span class="st">"fp"</span>,<span class="st">"fn"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">WT_vs_KO_coords</span><span class="op">$</span><span class="va">sum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">WT_vs_KO_coords</span><span class="op">$</span><span class="va">tp</span> <span class="op">+</span> <span class="va">WT_vs_KO_coords</span><span class="op">$</span><span class="va">fp</span><span class="op">)</span></span>
<span><span class="va">WT_vs_KO_coords</span><span class="op">$</span><span class="va">ppv</span> <span class="op">&lt;-</span> <span class="va">WT_vs_KO_coords</span><span class="op">$</span><span class="va">ppv</span> <span class="op">*</span> <span class="fl">20</span></span>
<span><span class="va">WT_vs_KO_coords</span> <span class="op">&lt;-</span> <span class="va">WT_vs_KO_coords</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">WT_vs_KO_coords</span><span class="op">$</span><span class="va">sum</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">WT_vs_KO_coords</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">threshold</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">sum</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"#009E73"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">ppv</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"#D55E00"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span></span>
<span>    <span class="co"># Features of the first axis</span></span>
<span>    name <span class="op">=</span> <span class="st">"Log2(#predicted modified)"</span>,</span>
<span>    <span class="co"># Add a second axis and specify its features</span></span>
<span>    sec.axis <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/sec_axis.html" class="external-link">sec_axis</a></span><span class="op">(</span><span class="op">~</span><span class="va">.</span><span class="op">*</span><span class="fl">5</span>, name <span class="op">=</span> <span class="st">"PPV (miCLIP)"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span>  </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"JACUSA2 NMF Score"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>,      </span>
<span>    axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"#009E73"</span>, size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    axis.title.y.right <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"#D55E00"</span>, size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">0.9</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel</a></span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="JACUSA2helper-nanopore_files/figure-html/unnamed-chunk-42-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-boulias_identification_2019" class="csl-entry">
Boulias, Konstantinos, Diana Toczydłowska-Socha, Ben R. Hawley, Noa
Liberman, Ken Takashima, Sara Zaccara, Théo Guez, et al. 2019.
<span>“Identification of the <span class="nocase">m6Am</span>
<span>Methyltransferase</span> <span>PCIF1</span> <span>Reveals</span>
the <span>Location</span> and <span>Functions</span> of <span class="nocase">m6Am</span> in the <span>Transcriptome</span>.”</span>
<em>Molecular Cell</em> 75 (3): 631–643.e8. <a href="https://doi.org/10.1016/j.molcel.2019.06.006" class="external-link">https://doi.org/10.1016/j.molcel.2019.06.006</a>.
</div>
<div id="ref-koh_atlas_2019" class="csl-entry">
Koh, Casslynn W. Q., Yeek Teck Goh, and W. S. Sho Goh. 2019.
<span>“Atlas of Quantitative Single-Base-Resolution
<span>N6</span>-Methyl-Adenine Methylomes.”</span> <em>Nature
Communications</em> 10 (1): 5636. <a href="https://doi.org/10.1038/s41467-019-13561-z" class="external-link">https://doi.org/10.1038/s41467-019-13561-z</a>.
</div>
<div id="ref-kortel_deep_2021" class="csl-entry">
Körtel, Nadine, Cornelia Rücklé, You Zhou, Anke Busch, Peter Hoch-Kraft,
F X Reymond Sutandy, Jacob Haase, et al. 2021. <span>“Deep and Accurate
Detection of <span class="nocase">m6A</span> <span>RNA</span>
Modifications Using <span class="nocase">miCLIP2</span> and <span class="nocase">m6Aboost</span> Machine Learning.”</span> <em>Nucleic
Acids Research</em> 49 (16): e92–92. <a href="https://doi.org/10.1093/nar/gkab485" class="external-link">https://doi.org/10.1093/nar/gkab485</a>.
</div>
<div id="ref-Piechotta2021" class="csl-entry">
Piechotta, Michael, Qi Wang, Janine Altmüller, and Christoph Dieterich.
2021. <span>“<span>RNA</span> Modification Mapping with
<span>JACUSA</span>2,”</span> July. <a href="https://doi.org/10.1101/2021.07.02.450888" class="external-link">https://doi.org/10.1101/2021.07.02.450888</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Michael Piechotta, <a href="http://dieterichlab.org/" class="external-link">Dieterich Lab</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
