<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>JACUSA2helper • JACUSA2helper</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="JACUSA2helper">
<meta property="og:description" content="JACUSA2helper">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">JACUSA2helper</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.99-10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/JACUSA2helper.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/JACUSA2helper-figure1e.html">JACUSA2helper Figure 1E</a>
    </li>
    <li>
      <a href="../articles/JACUSA2helper-figure1f.html">JACUSA2helper Figure 1F</a>
    </li>
    <li>
      <a href="../articles/JACUSA2helper-meta-condtions.html">JACUSA2helper meta conditions</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dieterich-lab/JACUSA2helper/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="JACUSA2helper_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>JACUSA2helper</h1>
                        <h4 class="author">Michael Piechotta</h4>
            
            <h4 class="date">2021-07-12</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/dieterich-lab/JACUSA2helper/blob/master/vignettes/JACUSA2helper.Rmd"><code>vignettes/JACUSA2helper.Rmd</code></a></small>
      <div class="hidden name"><code>JACUSA2helper.Rmd</code></div>

    </div>

    
    
<p>A typical workflow for analyzing <a href="https://github.com/dieterich-lab/JACUSA2">JACUSA2</a> output file(s) with <a href="https://github.com/dieterich-lab/JACUSA2helper">JACUSA2helper</a> consists of:</p>
<ul>
<li>Read JACUSA2 output file(s) into result object.</li>
<li>Add optional data, e.g.: <code><a href="../reference/base_sub.html">base_sub()</a></code> - add base substitution.</li>
<li>Filter result object by some criteria, e.g.: <code><a href="https://dplyr.tidyverse.org/reference/filter.html">dplyr::filter(All(cov$cond1 &gt;= 10))</a></code> - retain sites with read coverage &gt;= 10, and use <code><a href="../reference/All.html">All()</a></code> or <code>Any</code> on structured columns (conditions and replicates).</li>
<li>Plot filtered result object.</li>
</ul>
<p>JACUSA2helper supports the analysis of the following methods from JACUSA2: <em>call-{1,2}</em>, <em>pileup</em>, <em>rt-arrest</em>, and <em>lrt-arrest</em>(experimental). Check the <a href="https://github.com/dieterich-lab/JACUSA2/blob/master/manual/manual.pdf">JACUSA2 manual</a> for more details.</p>
<p>The main data structure is the result object that is implemented by following the tidy data approach from <span class="citation">Wickham (<a href="#ref-Wickham2014" role="doc-biblioref">2014</a>)</span> to feature easy interaction with <a href="https://github.com/tidyverse/dplyr">dplyr</a> and <a href="https://github.com/tidyverse/ggplot2">ggplot2</a>.</p>
<p>In the following, we will focus on single pairwise comparisons of two conditions of one output file. If you want to analyze multiple files with similar pairwise comparisons, read <code>vignette("JACUSA2helper meta conditions")</code>.</p>
<div id="multi-site-and-new-features" class="section level2">
<h2 class="hasAnchor">
<a href="#multi-site-and-new-features" class="anchor"></a>Multi site and new features</h2>
<p>While previously in JACUSA<em>1</em>, a site could be uniquely identified as one line in the output by the coordinate columns: ‘contig’, ‘start’, ‘end’, and ‘strand’, JACUSA<em>2</em> features more complex data structures to store new features such as arrest positions of arrest events, variant tagging, and INDEL counts. Therefore, sites can now cover multiple lines of output - check Section “Introduction” of <a href="https://github.com/dieterich-lab/JACUSA2/blob/master/manual/manual.pdf">JACUSA2 manual</a> for details.</p>
</div>
<div id="variant-calling---data-rdd-rna-dna-differences" class="section level2">
<h2 class="hasAnchor">
<a href="#variant-calling---data-rdd-rna-dna-differences" class="anchor"></a>Variant calling - Data: rdd (RNA-DNA-differences)</h2>
<p>To introduce the basic verbs for manipulating JAUCSA2 output, we’ll use <code>JACUSA2helper:rdd</code>. This data set contains a subset from <span class="citation">Piechotta et al. (<a href="#ref-Piechotta2017" role="doc-biblioref">2017</a>)</span> and is documented in <code><a href="../reference/rdd.html">?rdd</a></code>. In brief, the sample data set consists of 10.000 sites of 1x DNA and 2x stranded RNA sequencing libraries that allow to define RNA-DNA-differences (RDDs).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dieterich-lab/JACUSA2helper">JACUSA2helper</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">rdd</span><span class="op">)</span></code></pre></div>
<p>Note that <code>rdd</code> is a tibble - check <a href="http://tibble.tidyverse.org">tibble</a> for more details.</p>
<p>Use dplyr to manipulate the result object <code>rdd</code>. Check details on <a href="https://github.com/tidyverse/dplyr">dplyr</a> and read <code>vignette("dplyr")</code>.</p>
<div id="structured-column" class="section level3">
<h3 class="hasAnchor">
<a href="#structured-column" class="anchor"></a>Structured column</h3>
<p>JACUSA2 features structured columns (nested tibbles) where condition and replicate specific data is stored, e.g.: bases, cov, etc.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">rdd</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"id"</span>, <span class="st">"bases"</span>, <span class="st">"cov"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt; Classes 'tbl_df', 'tbl' and 'data.frame':    94 obs. of  3 variables:</span>
<span class="co">#&gt;  $ id   : chr  "GL000194.1:11096:-" "X:135997095:+" "1:100709076:-" "1:142790269:-" ...</span>
<span class="co">#&gt;  $ bases:Classes 'tbl_df', 'tbl' and 'data.frame':   94 obs. of  2 variables:</span>
<span class="co">#&gt;   ..$ cond1:Classes 'tbl_df', 'tbl' and 'data.frame':    94 obs. of  1 variable:</span>
<span class="co">#&gt;   .. ..$ rep1:Classes 'tbl_df', 'tbl' and 'data.frame':  94 obs. of  4 variables:</span>
<span class="co">#&gt;   .. .. ..$ A: num  0 16 36 0 25 29 41 0 0 2 ...</span>
<span class="co">#&gt;   .. .. ..$ C: num  0 0 0 0 0 0 0 4 19 0 ...</span>
<span class="co">#&gt;   .. .. ..$ G: num  5 0 0 0 0 0 1 0 0 0 ...</span>
<span class="co">#&gt;   .. .. ..$ T: num  0 0 0 9 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;   ..$ cond2:Classes 'tbl_df', 'tbl' and 'data.frame':    94 obs. of  2 variables:</span>
<span class="co">#&gt;   .. ..$ rep1:Classes 'tbl_df', 'tbl' and 'data.frame':  94 obs. of  4 variables:</span>
<span class="co">#&gt;   .. .. ..$ A: num  9 0 1 0 0 0 1 4 2 1 ...</span>
<span class="co">#&gt;   .. .. ..$ C: num  0 0 0 5 0 0 0 1 1 2 ...</span>
<span class="co">#&gt;   .. .. ..$ G: num  1 3 3 0 9 7 5 0 0 0 ...</span>
<span class="co">#&gt;   .. .. ..$ T: num  0 0 0 1 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;   .. ..$ rep2:Classes 'tbl_df', 'tbl' and 'data.frame':  94 obs. of  4 variables:</span>
<span class="co">#&gt;   .. .. ..$ A: num  5 1 0 0 5 1 4 5 4 1 ...</span>
<span class="co">#&gt;   .. .. ..$ C: num  0 0 0 3 0 0 0 0 0 3 ...</span>
<span class="co">#&gt;   .. .. ..$ G: num  0 2 3 0 10 10 3 0 0 0 ...</span>
<span class="co">#&gt;   .. .. ..$ T: num  0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;  $ cov  :Classes 'tbl_df', 'tbl' and 'data.frame':   94 obs. of  2 variables:</span>
<span class="co">#&gt;   ..$ cond1:Classes 'tbl_df', 'tbl' and 'data.frame':    94 obs. of  1 variable:</span>
<span class="co">#&gt;   .. ..$ rep1: num  5 16 36 9 25 29 42 4 19 2 ...</span>
<span class="co">#&gt;   ..$ cond2:Classes 'tbl_df', 'tbl' and 'data.frame':    94 obs. of  2 variables:</span>
<span class="co">#&gt;   .. ..$ rep1: num  10 3 4 6 9 7 6 5 3 3 ...</span>
<span class="co">#&gt;   .. ..$ rep2: num  5 3 3 3 15 11 7 5 4 4 ...</span></code></pre></div>
<p>Check structure of column with:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">rdd</span><span class="op">$</span><span class="va">cov</span><span class="op">)</span>
<span class="co">#&gt; [1] "cond1" "cond2"</span></code></pre></div>
<p>To access specific coverage information for condition 2 use:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">rdd</span><span class="op">$</span><span class="va">cov</span><span class="op">$</span><span class="va">cond2</span><span class="op">)</span>
<span class="co">#&gt; Classes 'tbl_df', 'tbl' and 'data.frame':    94 obs. of  2 variables:</span>
<span class="co">#&gt;  $ rep1: num  10 3 4 6 9 7 6 5 3 3 ...</span>
<span class="co">#&gt;  $ rep2: num  5 3 3 3 15 11 7 5 4 4 ...</span></code></pre></div>
<p>In summary, all structured/nested columns hold: * data of a condition “cond<em>I</em>” on the first level, and * data of a replicate “rep<em>J</em>” on the second level.</p>
<p>Coverage information for condition 2 and replicate 1 can be accessed with:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">rdd</span><span class="op">$</span><span class="va">cov</span><span class="op">$</span><span class="va">cond2</span><span class="op">$</span><span class="va">rep1</span><span class="op">)</span>
<span class="co">#&gt;  num [1:94] 10 3 4 6 9 7 6 5 3 3 ...</span></code></pre></div>
<div id="wrapper-for-lapply-and-mapply" class="section level4">
<h4 class="hasAnchor">
<a href="#wrapper-for-lapply-and-mapply" class="anchor"></a>Wrapper for <code>lapply()</code> and <code>mapply()</code>
</h4>
<p>For accessing and manipulating structured columns there are JACUSA2 specific wrappers for <code><a href="https://rdrr.io/r/base/lapply.html">lapply()</a></code> and <code><a href="https://rdrr.io/r/base/mapply.html">mapply()</a></code>: * <code><a href="../reference/lapply_cond.html">lapply_cond(x, f)</a></code> operates on condition data (list of replicate data), * <code><a href="../reference/lapply_repl.html">lapply_repl(x, f)</a></code> operates on replicate data of each condition, and * <code><a href="../reference/mapply_repl.html">mapply_repl(f, ...)</a></code> operates on replicate data, takes multiple inputs.</p>
<p>To get the total coverage for each condition use:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/lapply_cond.html">lapply_cond</a></span><span class="op">(</span><span class="va">rdd</span><span class="op">$</span><span class="va">cov</span>, <span class="va">rowSums</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 94 x 2</span>
<span class="co">#&gt;   cond1 cond2</span>
<span class="co">#&gt;   &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1     5    15</span>
<span class="co">#&gt; 2    16     6</span>
<span class="co">#&gt; 3    36     7</span>
<span class="co">#&gt; 4     9     9</span>
<span class="co">#&gt; # … with 90 more rows</span></code></pre></div>
<p>The following gives you the observed base calls:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">base_calls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lapply_repl.html">lapply_repl</a></span><span class="op">(</span><span class="va">rdd</span><span class="op">$</span><span class="va">bases</span>, <span class="va">base_call</span><span class="op">)</span>
<span class="va">base_calls</span><span class="op">$</span><span class="va">cond2</span>
<span class="co">#&gt; # A tibble: 94 x 2</span>
<span class="co">#&gt;   rep1  rep2 </span>
<span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;</span>
<span class="co">#&gt; 1 AG    A    </span>
<span class="co">#&gt; 2 G     AG   </span>
<span class="co">#&gt; 3 AG    G    </span>
<span class="co">#&gt; 4 CT    C    </span>
<span class="co">#&gt; # … with 90 more rows</span></code></pre></div>
</div>
</div>
<div id="filtering-with-dplyrfilter" class="section level3">
<h3 class="hasAnchor">
<a href="#filtering-with-dplyrfilter" class="anchor"></a>Filtering with <code>dplyr::filter()</code>
</h3>
<p>dplyr and <code>%&gt;%</code> from magrittr allow to formulate compact analysis pipelines.</p>
<p>In a first step, we will retain sites with score &gt;= 2.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># before filtering</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">rdd</span><span class="op">)</span>
<span class="co">#&gt; [1] 94 15</span>
<span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">rdd</span>, <span class="va">score</span> <span class="op">&gt;=</span> <span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"After filtering"</span><span class="op">)</span>
<span class="co">#&gt; [1] "After filtering"</span>
<span class="co"># after filtering</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span>
<span class="co">#&gt; [1] 84 15</span></code></pre></div>
<p>In order to formulate complicate statements with <code><a href="https://dplyr.tidyverse.org/reference/filter.html">dplyr::filter()</a></code> featuring arbitrary combinations of conditions and/or replicates use the following convenience functions: * <code><a href="../reference/All.html">All()</a></code> and * <code><a href="../reference/Any.html">Any()</a></code>.</p>
<p>The following statement retains sites where all replicates of condition 1 have coverage &gt;= 50 and at least one replicate of condition 2 has 10 reads:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rdd</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="../reference/All.html">All</a></span><span class="op">(</span><span class="va">cov</span><span class="op">$</span><span class="va">cond1</span> <span class="op">&gt;=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="../reference/Any.html">Any</a></span><span class="op">(</span><span class="va">cov</span><span class="op">$</span><span class="va">cond2</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">id</span>, <span class="va">cov</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 1 x 2</span>
<span class="co">#&gt;   id            cov$cond1$rep1 $cond2$rep1 $$rep2</span>
<span class="co">#&gt;   &lt;chr&gt;                  &lt;dbl&gt;       &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt; 1 16:30931809:-             60          11      5</span></code></pre></div>
<p>In the following, we will:</p>
<ul>
<li>remove sites with score &lt; 2,</li>
<li>retain sites with coverage &gt;= 10 for all replicates,</li>
<li>remove sites with &gt; 2 observed bases (excluding reference base),</li>
<li>apply a filter that retains only robust sites (RNA editing must be present in all/both replicates),</li>
<li>Finally, we compute the base substitution (here RNA editing sites against FASTA reference) per replicate.</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">filtered</span> <span class="op">&lt;-</span> <span class="va">rdd</span>
<span class="va">filtered</span><span class="op">[[</span><span class="st">"bc"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lapply_cond.html">lapply_cond</a></span><span class="op">(</span><span class="va">filtered</span><span class="op">$</span><span class="va">bases</span>, <span class="kw">function</span><span class="op">(</span><span class="va">b</span><span class="op">)</span> <span class="op">{</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html">Reduce</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="va">b</span><span class="op">)</span> <span class="op">}</span> <span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html">Reduce</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="va">.</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/base_count.html">base_count</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">filtered</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">score</span> <span class="op">&gt;=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="../reference/All.html">All</a></span><span class="op">(</span><span class="va">cov</span><span class="op">$</span><span class="va">cond1</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="../reference/All.html">All</a></span><span class="op">(</span><span class="va">cov</span><span class="op">$</span><span class="va">cond2</span> <span class="op">&gt;</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">bc</span> <span class="op">&lt;=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="../reference/robust.html">robust</a></span><span class="op">(</span><span class="va">bases</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 3 x 16</span>
<span class="co">#&gt;   id     contig  start    end name  score strand bases$cond1$rep… bases$cond1$$C</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;   &lt;int&gt;  &lt;int&gt; &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;             &lt;dbl&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt; 1 14:10… 14     1.01e8 1.01e8 call…  12.2 -                    47              0</span>
<span class="co">#&gt; 2 19:48… 19     4.90e7 4.90e7 call…  11.3 +                     0             39</span>
<span class="co">#&gt; 3 21:11… 21     1.12e7 1.12e7 call…  16.2 +                     0              0</span>
<span class="co">#&gt; # … with 8 more variables: cov &lt;tibble[,2]&gt;, backtrack2 &lt;chr&gt;, resetP &lt;chr&gt;,</span>
<span class="co">#&gt; #   backtrackP &lt;chr&gt;, info &lt;chr&gt;, filter &lt;chr&gt;, ref &lt;chr&gt;, bc &lt;int&gt;</span>

<span class="co"># sum base call counts of condition / RNA replicates</span>
<span class="va">rna_bases</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html">Reduce</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="va">filtered</span><span class="op">$</span><span class="va">bases</span><span class="op">$</span><span class="va">cond2</span><span class="op">)</span>

<span class="co"># we don't need lapply_repl, because we don't operate on all replicate from all</span>
<span class="co"># conditions - only condition 2 / RNA</span>
<span class="va">ref2rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/base_sub.html">base_sub</a></span><span class="op">(</span><span class="va">rna_bases</span>, <span class="va">filtered</span><span class="op">$</span><span class="va">ref</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">ref2rna</span><span class="op">)</span>
<span class="co">#&gt; ref2rna</span>
<span class="co">#&gt; A-&gt;C A-&gt;G A-&gt;T C-&gt;A C-&gt;G C-&gt;T G-&gt;A G-&gt;C G-&gt;T T-&gt;A T-&gt;C </span>
<span class="co">#&gt;    4   51    5   12    1    6    3    1    1    2    8</span></code></pre></div>
<p>Instead of using DNA information from the reference FASTA sequence via column “ref”, we could use the actual data from condition 1 / DNA.</p>
<p>Robust requires observations (here base calls) to be present in all replicates of at least condition.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cond1_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/base_call.html">base_call</a></span><span class="op">(</span><span class="va">filtered</span><span class="op">$</span><span class="va">bases</span><span class="op">$</span><span class="va">cond1</span><span class="op">$</span><span class="va">rep1</span><span class="op">)</span>
<span class="va">dna2rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/base_sub.html">base_sub</a></span><span class="op">(</span><span class="va">rna_bases</span>, <span class="va">filtered</span><span class="op">$</span><span class="va">ref</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">dna2rna</span><span class="op">)</span>
<span class="co">#&gt; dna2rna</span>
<span class="co">#&gt; A-&gt;C A-&gt;G A-&gt;T C-&gt;A C-&gt;G C-&gt;T G-&gt;A G-&gt;C G-&gt;T T-&gt;A T-&gt;C </span>
<span class="co">#&gt;    4   51    5   12    1    6    3    1    1    2    8</span></code></pre></div>
<p>Using different sources of DNA to describe RRDs via <code>ref2rna</code> and <code>dna2rna</code> we can conclude that there are no polymorphic positions in <code>filtered</code> because <code>ref2rna</code> and <code>dna2rna</code> are identical.</p>
</div>
<div id="plot-base-substitutions" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-base-substitutions" class="anchor"></a>Plot base substitutions</h3>
<p>We can now plot the site specific distribution of base substitutions using <a href="https://github.com/tidyverse/ggplot2">ggplot2</a> with:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>base_sub <span class="op">=</span> <span class="va">dna2rna</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co"># ggplot requires a data frame</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">base_sub</span>, fill <span class="op">=</span> <span class="va">base_sub</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"Base substitution"</span><span class="op">)</span>  <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>
   legend.position <span class="op">=</span> <span class="st">"none"</span>, 
   axis.text.x <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, vjust <span class="op">=</span> <span class="fl">0.5</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<p><img src="JACUSA2helper_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
</div>
<div id="plot-coverage-distribution" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-coverage-distribution" class="anchor"></a>Plot coverage distribution</h3>
<p>To compare the coverage distribution of each data sample, we create a simple plot of the empirical coverage distribution, where we map cond(ition) to <code>colour</code> and repl(icate) to <code>linetype</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/gather_repl.html">gather_repl</a></span><span class="op">(</span><span class="va">result</span><span class="op">$</span><span class="va">id</span>, <span class="va">result</span><span class="op">$</span><span class="va">cov</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">value</span>, colour <span class="op">=</span> <span class="va">cond</span>, linetype <span class="op">=</span> <span class="va">repl</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/stat_ecdf.html">stat_ecdf</a></span><span class="op">(</span>geom <span class="op">=</span> <span class="st">"step"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"Coverage"</span><span class="op">)</span></code></pre></div>
<p><img src="JACUSA2helper_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>In order to improve the preliminary plot, we need to modify the following aspects:</p>
<ul>
<li>Merge legends,</li>
<li>add descriptive plot labels (x = Coverage, y = Density),</li>
<li>add descriptive data description - What is cond(ition) 1 and 2, and finally</li>
<li>logarithmize cov(erage).</li>
</ul>
<p>First, a data description is added to the result object:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gather_repl.html">gather_repl</a></span><span class="op">(</span><span class="va">rdd</span><span class="op">$</span><span class="va">id</span>, <span class="va">rdd</span><span class="op">$</span><span class="va">cov</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>
    data_desc <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html">case_when</a></span><span class="op">(</span>
      <span class="va">cond</span> <span class="op">==</span> <span class="st">"cond1"</span> <span class="op">~</span> <span class="st">"DNA"</span>,
      <span class="va">cond</span> <span class="op">==</span> <span class="st">"cond2"</span> <span class="op">&amp;</span> <span class="va">repl</span> <span class="op">==</span> <span class="st">"rep1"</span> <span class="op">~</span> <span class="st">"RNA 1"</span>,
      <span class="va">cond</span> <span class="op">==</span> <span class="st">"cond2"</span> <span class="op">&amp;</span> <span class="va">repl</span> <span class="op">==</span> <span class="st">"rep2"</span> <span class="op">~</span> <span class="st">"RNA 2"</span>
    <span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<p>Supplementary data <code>limits</code> and <code>labels</code> are created to enable sleek legend by relating cond(ition) and repl(icate) to data description. They will be mapped to <code>colour</code> and <code>linetype</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># grouping enables nice legends in plot</span>
<span class="va">result</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">interaction</span>, <span class="va">result</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"cond"</span>, <span class="st">"repl"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="co"># relate group={cond(ition), repl(icat)}, and nice label</span>
<span class="va">meta_desc</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html">distinct</a></span><span class="op">(</span><span class="va">result</span>, <span class="va">cond</span>, <span class="va">repl</span>, <span class="va">group</span>, <span class="va">data_desc</span><span class="op">)</span>
<span class="co"># map group values to nice labels (data_desc)</span>
<span class="va">limits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="va">meta_desc</span><span class="op">[[</span><span class="st">"group"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="va">labels</span> <span class="op">&lt;-</span> <span class="va">meta_desc</span><span class="op">[[</span><span class="st">"data_desc"</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p>Finally, we combine the previous snippets and arrive at the final plot:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">name</span> <span class="op">&lt;-</span> <span class="st">"Data description"</span>
<span class="va">result</span> <span class="op">%&gt;%</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">value</span>, colour <span class="op">=</span> <span class="va">group</span>, linetype <span class="op">=</span> <span class="va">group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="co"># map values for cond to colour and use nice labels</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>
    name <span class="op">=</span> <span class="va">name</span>,
    labels <span class="op">=</span> <span class="va">labels</span>,
    limits <span class="op">=</span> <span class="va">limits</span>,
    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">meta_desc</span><span class="op">[[</span><span class="st">"cond"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span><span class="op">(</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="co"># map values for repl to colour and use nice labels</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_linetype_manual</a></span><span class="op">(</span>
    name <span class="op">=</span> <span class="va">name</span>,
    labels <span class="op">=</span> <span class="va">labels</span>,
    limits <span class="op">=</span> <span class="va">limits</span>,
    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">meta_desc</span><span class="op">[[</span><span class="st">"repl"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span><span class="op">(</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/stat_ecdf.html">stat_ecdf</a></span><span class="op">(</span>geom <span class="op">=</span> <span class="st">"step"</span><span class="op">)</span> <span class="op">+</span>
  <span class="co"># theming and xy labels</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"Density"</span><span class="op">)</span> <span class="op">+</span>
  <span class="co"># logarithmize and nice x-axis</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span>
    breaks <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/trans_breaks.html">trans_breaks</a></span><span class="op">(</span><span class="st">"log10"</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fl">10</span><span class="op">^</span><span class="va">x</span><span class="op">)</span>,
    labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/trans_format.html">trans_format</a></span><span class="op">(</span><span class="st">"log10"</span>, <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/label_parse.html">math_format</a></span><span class="op">(</span><span class="fl">10</span><span class="op">^</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"Coverage"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span>  </code></pre></div>
<p><img src="JACUSA2helper_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>From the plot, we can deduce that RNA samples have similar coverage distribution and are higher covered than the DNA data sample. This code serves as a blueprint for other plots, e.g.:</p>
<ul>
<li>empirical cumulative distribution of arrest rate or</li>
<li>number of observed bases.</li>
</ul>
</div>
</div>
<div id="arrest-events" class="section level2">
<h2 class="hasAnchor">
<a href="#arrest-events" class="anchor"></a>Arrest events</h2>
<p>In <span class="citation">Zhou et al. (<a href="#ref-Zhou2018" role="doc-biblioref">2018</a>)</span> the authors map RNA modification of pseudouridine (<span class="math inline">\(\Psi\)</span>) by chemically modifying pseudouridines with carbodiimide (+CMC) and detecting arrest events that are induced by reverse transcription stops in high-throughput sequencing under 3 different conditions: HIVRT, SIIIRTMn, and SIIIRTMg. The result of JACUSA2 are available by using <code><a href="https://rdrr.io/r/utils/data.html">data()</a></code>. Read <a href="https://github.com/dieterich-lab/JACUSA2/blob/master/manual/manual.pdf">JACUSA2 manual</a> for details on how data has been processed from the original publication by <span class="citation">Zhou et al. (<a href="#ref-Zhou2018" role="doc-biblioref">2018</a>)</span>. In brief, the data has been filtered to contain the following rRNAs: RNA18SN5, RNA28SN5, and RNA5-8SN5.</p>
<p>In the following, we will be looking at the results for <code><a href="https://rdrr.io/r/utils/data.html">data(HIVRT_rt_arrest)</a></code>. By default JACUSA2 output from <em>rt-arrest</em> will contain the following structured/nested columns:</p>
<ul>
<li>
<code>arrest</code> and <code>through</code>
</li>
<li>
<code>bases</code> = <code>arrest</code> + <code>through</code>: total base counts.</li>
<li><code>arrest_rate</code></li>
<li>
<code>cov</code> holds coverage for total base counts.</li>
</ul>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"HIVRT_rt_arrest"</span><span class="op">)</span>
<span class="co"># column names of arrest data</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">HIVRT_rt_arrest</span><span class="op">)</span>
<span class="co">#&gt;  [1] "id"           "contig"       "start"        "end"          "name"        </span>
<span class="co">#&gt;  [6] "pvalue"       "strand"       "arrest"       "through"      "bases"       </span>
<span class="co">#&gt; [11] "arrest_rate"  "cov"          "arrest_score" "backtrackP"   "resetP"      </span>
<span class="co">#&gt; [16] "info"         "filter"       "ref"</span></code></pre></div>
<p>We investigate the strand specific coverage distribution of reads and discover that more reads are mapped to the “+” strand. The other mappings are probably artefacts. But this discussion is beyond the scope of this vignette.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/gather_repl.html">gather_repl</a></span><span class="op">(</span><span class="va">HIVRT_rt_arrest</span><span class="op">$</span><span class="va">id</span>, <span class="va">HIVRT_rt_arrest</span><span class="op">$</span><span class="va">cov</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co"># extract coverages</span>
<span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">inner_join</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">HIVRT_rt_arrest</span>, <span class="va">id</span>, <span class="va">contig</span>, <span class="va">strand</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">"id"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co"># add contig and strand</span>
<span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">value</span>, colour <span class="op">=</span> <span class="va">strand</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span>
    breaks <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/trans_breaks.html">trans_breaks</a></span><span class="op">(</span><span class="st">"log10"</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fl">10</span><span class="op">^</span><span class="va">x</span><span class="op">)</span>,
    labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/trans_format.html">trans_format</a></span><span class="op">(</span><span class="st">"log10"</span>, <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/label_parse.html">math_format</a></span><span class="op">(</span><span class="fl">10</span><span class="op">^</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">+</span>
<span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"Coverage"</span><span class="op">)</span> <span class="op">+</span>
<span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span> 
<span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span><span class="va">contig</span> <span class="op">~</span> <span class="va">.</span><span class="op">)</span></code></pre></div>
<p><img src="JACUSA2helper_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>We filter by pvalue, coverage, strand, and only retain robust arrest events:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">HIVRT_rt_arrest</span><span class="op">)</span>
<span class="co">#&gt; [1] 12110    18</span>
<span class="va">filtered</span> <span class="op">&lt;-</span> <span class="va">HIVRT_rt_arrest</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">strand</span> <span class="op">==</span> <span class="st">"+"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">pvalue</span> <span class="op">&lt;=</span> <span class="fl">0.01</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="../reference/robust.html">robust</a></span><span class="op">(</span><span class="va">arrest</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">filtered</span><span class="op">)</span>
<span class="co">#&gt; [1] 1204   18</span></code></pre></div>
<p>We use <code><a href="../reference/gather_repl.html">gather_repl()</a></code> function to transform the filtered result object and plot the arrest rate from each condition in a scatter plot.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">filtered</span> <span class="op">%&gt;%</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">arrest_rate</span><span class="op">$</span><span class="va">cond1</span><span class="op">$</span><span class="va">rep1</span><span class="op">)</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">arrest_rate</span><span class="op">$</span><span class="va">cond2</span><span class="op">$</span><span class="va">rep1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>  
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span><span class="va">.</span> <span class="op">~</span> <span class="va">contig</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_abline</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"log2(+CMC arrest rate)"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"log2(-CMC arrest rate)"</span><span class="op">)</span></code></pre></div>
<p><img src="JACUSA2helper_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>From the above plot, we can deduce that the arrest rate is higher in the +CMC condition for all rRNAs when compared against -CMC.</p>
<p>If you want to look simultaneously at all conditions (HIVRT, SIIIRTMn, and SIIIRTMg), checkout <code>vignette("JACUSA2helper-meta-conditions")</code>.</p>
</div>
<div id="stranded-data" class="section level2">
<h2 class="hasAnchor">
<a href="#stranded-data" class="anchor"></a>Stranded data</h2>
<p>When working with stranded RNA-Seq data, inverting base calls is not necessary because JACUSA2 will automatically invert Single End (SE) and Paired End (PE) depending on the provided library type option “-P” UNSTRANDED|FR_FIRSTSTRAND|RF_SECONDSTRAND".</p>
</div>
<div id="inputoutput" class="section level1">
<h1 class="hasAnchor">
<a href="#inputoutput" class="anchor"></a>Input/Output</h1>
<p>There are two functions to read JACUSA2 output <code><a href="../reference/read_result.html">read_result(file)</a></code> and <code><a href="../reference/read_results.html">read_results(files, meta_conds)</a></code> which is explained in <code>vignette("JACUSA2helper-meta-conditions")</code>.</p>
<p>Use <code>result &lt;- read_result("JACUSA2.out")</code> to read and create a JACUSA2 result object from a JACUSA2 output file <code>JACUSA2.out</code>. By default, variables stored in the <strong>info</strong> column will be processed and unpacked making them available for further manipulation.</p>
<p>Summary of input/output methods:</p>
<ul>
<li>
<code><a href="../reference/read_result.html">read_result()</a></code> Reads and unpacks a JACUSA2 result file and creates a result object.</li>
<li>
<code><a href="../reference/read_results.html">read_results()</a></code> Allows to combine multiple result files and distinguish them with meta conditions.</li>
<li>
<code><a href="../reference/write_bedGraph.html">write_bedGraph()</a></code> Writes a vector of values as bedGraph file.</li>
</ul>
</div>
<div id="general-function-layout" class="section level1">
<h1 class="hasAnchor">
<a href="#general-function-layout" class="anchor"></a>General function layout</h1>
<p>In the following the core functions of JACUSA2helper are presented. Check the respective help page, e.g.: <code>?bc_ratio</code> to get more details.</p>
<p><code>arrest_rate</code> Calculate arrest rate from arrest and through reads. <code>base_call</code> Calculates observed bases from base counts. <code>base_ratio</code> Calculates a ratio matrix from a base count matrix. <code>coverage</code> Calculates coverage from base counts. In most cases this will be called automatically in <code><a href="../reference/read_result.html">read_result()</a></code> <code>non_ref_ratio</code> Calculates non reference base ratio for base counts of some base type. <code>base_sub</code> Calculates base substitution from reference and base counts, e.g.: A-&gt;G. <code>sub_ratio</code> Calculates base substitution ratio from reference and observed bases, e.g.: 25% A-&gt;G.</p>
</div>
<div id="filter-result-object" class="section level1">
<h1 class="hasAnchor">
<a href="#filter-result-object" class="anchor"></a>Filter result object</h1>
<ul>
<li><code><a href="../reference/robust.html">robust()</a></code></li>
<li><code><a href="../reference/filter_artefact.html">filter_artefact()</a></code></li>
<li><code>All(), Any()</code></li>
</ul>
</div>
<div id="variant-tagging" class="section level1">
<h1 class="hasAnchor">
<a href="#variant-tagging" class="anchor"></a>Variant tagging</h1>
<p>Variant tagging allows to group data based on base substitutions on reads, called tags. Running JACUSA2 with <code>-B "A2G"</code> will tag all reads that have a “A-&gt;G” base substitution. Where available, each observations for each site (=contig, start, stop, strand) from tagged and all reads will be provided and distinguished by “tag=A2G” or "tag=*" in the info column, respectively. Use <code>result$tag &lt;- clean_tag(result$tag)</code> to convert base substitutions such as “A2G” to “A-&gt;G”.</p>
<!--
#```{r}
## load data
#data(APOBEC1)
#result <- APOBEC1
#barplot(table(result$tag), ylab="# of sites", main="Distribution of tagged sites")
#```
#We calculate the total base call count with `lapply_cond` and `Reduce("+")`.

#```{r}
## sum base call counts of conditions and replicates
#result$total_bases <- lapply_cond(
#  result$bases, function(bases) { Reduce("+", bases) } 
#) %>% Reduce("+", .)
#```


#```{r}
## mark "valid" sites 
#result <- result %>%
#  dplyr::mutate(
#    keep = (
#      # restrict to total reads
#      tag == "*" & 
#      # require less than 2 different base calls
#      base_count(total_bases, ref) <= 2 & 
#      # require 10 reads for each replicate of each condition
#      All(cov$cond1 >= 10) & All(cov$cond2 >= 10) &
#      score >= 2 &
#      robust(bases)
#    )
#  )
## raw data
## add structured columns: tagged_bases and not_tagged_bases
#expanded <- expand_tag(result)
#filtered <- dplyr::filter(expanded, keep == TRUE)
#```

#```{r}
## plot
#count <- c(nrow(result), nrow(expanded), nrow(filtered))
#b <- barplot(count, names.arg = c("raw", "expaded", "filtered"), ylim=c(0, max(count) + max(count) * 15 / 100))
#text(b, count + count * 10 / 100, count)
#```

#```{r} 
## add observed non reference base substitution, e.g.: A->G
#filtered[["base_sub"]] <- base_sub(filtered$total_bases, filtered$ref)
## ratio of non reference base calls
#filtered[["non_ref_ratio"]] <- lapply_repl(filtered$bases, function(cond) non_ref_ratio(cond, filtered$ref))
#
## tag specific base substitution for each condition
#filtered[["tagged_base_sub"]] <- lapply_cond(filtered$tagged_bases, function(bases) { Reduce("+", bases) %>% base_sub(filtered$ref)} )
#filtered[["tagged_non_ref_ratio"]] <- lapply_repl(filtered$tagged_bases, function(cond) non_ref_ratio(cond, filtered$ref))
#
#filtered[["not_tagged_base_sub"]] <- lapply_cond(filtered$not_tagged_bases, function(bases) { Reduce("+", bases) %>% #base_sub(filtered$ref)} )
#filtered[["not_tagged_non_ref_ratio"]] <- lapply_repl(filtered$not_tagged_bases, function(cond) non_ref_ratio(cond, filtered$ref))
#```
-->
<!--
# Experimental features
* *lrt-arrest*
-->
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-Piechotta2017">
<p>Piechotta, Michael, Emanuel Wyler, Uwe Ohler, Markus Landthaler, and Christoph Dieterich. 2017. “JACUSA: Site-Specific Identification of RNA Editing Events from Replicate Sequencing Data.” <em>BMC Bioinformatics</em> 18 (1). <a href="https://doi.org/10.1186/s12859-016-1432-8">https://doi.org/10.1186/s12859-016-1432-8</a>.</p>
</div>
<div id="ref-Wickham2014">
<p>Wickham, Hadley. 2014. “Tidy Data.” <em>The Journal of Statistical Software</em> 59 (10). <a href="http://www.jstatsoft.org/v59/i10/">http://www.jstatsoft.org/v59/i10/</a>.</p>
</div>
<div id="ref-Zhou2018">
<p>Zhou, Katherine I., Wesley C. Clark, David W. Pan, Matthew J. Eckwahl, Qing Dai, and Tao Pan. 2018. “Pseudouridines Have Context-Dependent Mutation and Stop Rates in High-Throughput Sequencing.” <em>RNA Biology</em> 15 (7): 892–900. <a href="https://doi.org/10.1080/15476286.2018.1462654">https://doi.org/10.1080/15476286.2018.1462654</a>.</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Michael Piechotta, <a href="http://dieterichlab.org/">Dieterich Lab</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
