<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysis of Nanopore HEK293 with JACUSA2helper • JACUSA2helper</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of Nanopore HEK293 with JACUSA2helper">
<meta property="og:description" content="JACUSA2helper">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">JACUSA2helper</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.9.9.9100</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/JACUSA2helper.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/JACUSA2helper-figure1e.html">Analysis of m6A mapping by MazF with JACUSA2helper</a>
    </li>
    <li>
      <a href="../articles/JACUSA2helper-figure1f.html">Analysis of DART-seq with JACUSA2helper</a>
    </li>
    <li>
      <a href="../articles/JACUSA2helper-meta-condtions.html">Introduction to meta conditions with JACUSA2helper</a>
    </li>
    <li>
      <a href="../articles/JACUSA2helper-nanopore.html">Analysis of Nanopore HEK293 with JACUSA2helper</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dieterich-lab/JACUSA2helper/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analysis of Nanopore HEK293 with JACUSA2helper</h1>
                        <h4 data-toc-skip class="author">Christoph Dieterich, Michael Piechotta</h4>
            
            <h4 data-toc-skip class="date">2022-02-02</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/dieterich-lab/JACUSA2helper/blob/HEAD/vignettes/JACUSA2helper-nanopore.Rmd" class="external-link"><code>vignettes/JACUSA2helper-nanopore.Rmd</code></a></small>
      <div class="hidden name"><code>JACUSA2helper-nanopore.Rmd</code></div>

    </div>

    
    
<p>In the following, the workflow for use case 3 from <span class="citation">(<a href="#ref-Piechotta20221" role="doc-biblioref"><strong>Piechotta20221?</strong></a>)</span> is presented. Data for <a href="https://zenodo.org/record/5930729#.YfmTDFvMIUG" class="external-link">use cases 1-3</a> can be downloaded to repeat the analysis.</p>
<div class="section level2">
<h2 id="read-and-filter-jacusa2-output">Read and filter JACUSA2 output<a class="anchor" aria-label="anchor" href="#read-and-filter-jacusa2-output"></a>
</h2>
<p>Use <code>read_result</code> to read JACUSA2 output and filter the data set.</p>
<p>To repeat the Nanopore analysis from <span class="citation"><a href="#ref-Piechotta2021" role="doc-biblioref">Piechotta et al.</a> (<a href="#ref-Piechotta2021" role="doc-biblioref">2021</a>)</span> download the following data: * <a href="https://zenodo.org/record/5930729/files/WT_vs_KO_call2_result.out.gz?download=1" class="external-link">WT vs. KO</a> and * <a href="https://zenodo.org/record/5930729/files/WT_vs_realIVT_v202_call2_result.out.gz?download=1" class="external-link">WT vs. IVT</a>. Depending on your bandwidth the Download might take some time. Make sure, you have approx. 20GB of main memory available.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># files to read</span>
<span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>
  <span class="st">"/storage/zenodo/use_case3/WT_vs_KO_call2_result.out.gz"</span>,
  <span class="st">"/storage/zenodo/use_case3/WT_vs_realIVT_v202_call2_result.out.gz"</span>
<span class="op">)</span>

<span class="co"># descriptions corresponding to files </span>
<span class="va">meta_conds</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>
  <span class="st">"WT_vs_IVT"</span>,
  <span class="st">"WT_vs_KO"</span>
<span class="op">)</span></code></pre></div>
<p>In the following, we read multiple JACUSA2 output files and filter sites: * retain sites with “A” in the reference, * retain sites on chromosome 1-22, and X, and * remove sites that are within homopolymers (JACUSA2 filter flag: “Y”).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># read and filter</span>
<span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">file</span>, <span class="va">meta_cond</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_result.html">read_result</a></span><span class="op">(</span><span class="va">file</span>, nThread <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span>
        <span class="va">ref</span> <span class="op">==</span> <span class="st">"A"</span>,
        <span class="op">!</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"Y"</span>, <span class="va">filter</span>, fixed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,
        <span class="va">seqnames</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span>, <span class="st">"X"</span><span class="op">)</span>
      <span class="op">)</span>
    <span class="va">result</span><span class="op">$</span><span class="va">meta_cond</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">meta_cond</span>, <span class="va">meta_conds</span><span class="op">)</span>
    
    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span>
  <span class="op">}</span>, <span class="va">files</span>, <span class="va">meta_conds</span>, SIMPLIFY <span class="op">=</span> <span class="cn">FALSE</span>, USE.NAMES <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># convert and concatenate GenomicRanges</span>
<span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu">as</span><span class="op">(</span><span class="va">results</span>, <span class="st">"GRangesList"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># filtered sites per file / meta condition</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">meta_cond</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; WT_vs_IVT  WT_vs_KO </span>
<span class="co">#&gt;   2015555   1503040</span></code></pre></div>
<div class="section level4">
<h4 id="read_results">
<code>read_results</code><a class="anchor" aria-label="anchor" href="#read_results"></a>
</h4>
<p>An other option is to read the data set via <code>read_results</code> - see <code>vignette(Introduction to meta conditions with JACUSA2helper)</code>. This approach will consume much more main memory. We recommend to read big data sets serially with <code>read_result</code> and combine them afterwards.</p>
</div>
<div class="section level3">
<h3 id="unpack-info-field-and-reduce-data-set">Unpack “info” field and reduce data set<a class="anchor" aria-label="anchor" href="#unpack-info-field-and-reduce-data-set"></a>
</h3>
<p>The “info” field contains meta information for sites, such as detailed INDEL statistics. To save memory, we manually unpack the “info” field and select the following keys: * “insertion_score” and * “deletion_score.” We will use these data during the model training.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">unpacked_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unpack_info.html">unpack_info</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">info</span>, cond_count <span class="op">=</span> <span class="fl">2</span>, keys<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"insertion_score"</span>, <span class="st">"deletion_score"</span><span class="op">)</span><span class="op">)</span>
<span class="fu">mcols</span><span class="op">(</span><span class="va">results</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">results</span><span class="op">)</span>, <span class="va">unpacked_info</span><span class="op">)</span></code></pre></div>
<p>We continue to reduce the data set by selecting only relevant columns and renaming “score” column to “call2_score” for consistency:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">mcols</span><span class="op">(</span><span class="va">results</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">results</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"meta_cond"</span>, <span class="st">"score"</span>, <span class="st">"insertion_score"</span>, <span class="st">"deletion_score"</span><span class="op">)</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">results</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">results</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="st">"score"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"call2_score"</span>

<span class="va">results</span>
<span class="co">#&gt; GRanges object with 3518595 ranges and 4 metadata columns:</span>
<span class="co">#&gt;             seqnames    ranges strand | meta_cond call2_score insertion_score</span>
<span class="co">#&gt;                &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;factor&gt;   &lt;numeric&gt;       &lt;numeric&gt;</span>
<span class="co">#&gt;         [1]        1     14411      - | WT_vs_IVT    0.182221        0.973777</span>
<span class="co">#&gt;         [2]        1     14415      - | WT_vs_IVT    0.200339              NA</span>
<span class="co">#&gt;         [3]        1     14416      - | WT_vs_IVT    0.279846              NA</span>
<span class="co">#&gt;         [4]        1     14418      - | WT_vs_IVT    0.312013        1.740745</span>
<span class="co">#&gt;         [5]        1     14419      - | WT_vs_IVT    1.263700        0.236266</span>
<span class="co">#&gt;         ...      ...       ...    ... .       ...         ...             ...</span>
<span class="co">#&gt;   [3518591]        X 155943748      + |  WT_vs_KO    1.251611       0.1979329</span>
<span class="co">#&gt;   [3518592]        X 155943750      + |  WT_vs_KO    0.124106       0.0101097</span>
<span class="co">#&gt;   [3518593]        X 155943754      + |  WT_vs_KO    0.936613       0.0560419</span>
<span class="co">#&gt;   [3518594]        X 155943755      + |  WT_vs_KO    0.256918       1.0942557</span>
<span class="co">#&gt;   [3518595]        X 155943761      + |  WT_vs_KO    0.280098       2.7659899</span>
<span class="co">#&gt;             deletion_score</span>
<span class="co">#&gt;                  &lt;numeric&gt;</span>
<span class="co">#&gt;         [1]    -0.00084858</span>
<span class="co">#&gt;         [2]     0.84031855</span>
<span class="co">#&gt;         [3]             NA</span>
<span class="co">#&gt;         [4]     0.01443832</span>
<span class="co">#&gt;         [5]     0.13743735</span>
<span class="co">#&gt;         ...            ...</span>
<span class="co">#&gt;   [3518591]             NA</span>
<span class="co">#&gt;   [3518592]      0.4474059</span>
<span class="co">#&gt;   [3518593]      0.0352544</span>
<span class="co">#&gt;   [3518594]      0.2855172</span>
<span class="co">#&gt;   [3518595]             NA</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 31 sequences from an unspecified genome; no seqlengths</span></code></pre></div>
<p><code>results</code> now consists of locations and scores for sites from all <code>files</code> and <code>meta_conds</code>.</p>
<div class="section level4">
<h4 id="indicator-variables">Indicator variables<a class="anchor" aria-label="anchor" href="#indicator-variables"></a>
</h4>
<p>We extend the coordinates of a site by 2nt in each direction:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">extended_sites</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extend.html">extend</a></span><span class="op">(</span><span class="va">results</span>, left<span class="op">=</span><span class="fl">2</span>, right<span class="op">=</span><span class="fl">2</span><span class="op">)</span>

<span class="va">unique_regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">extended_sites</span><span class="op">)</span></code></pre></div>
<p><code>unique_regions</code> consists now of 5nt wide coordinates around all sites with the site positioned in the middle (= position 3). We calculate the position of each site in overlapping <code>unique_regions</code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># compute overlap</span>
<span class="va">hits</span> <span class="op">&lt;-</span> <span class="fu">findOverlaps</span><span class="op">(</span><span class="va">results</span>, <span class="va">unique_regions</span><span class="op">)</span>
<span class="co"># corresponding overlap in ...</span>
<span class="va">overlap_regions</span> <span class="op">&lt;-</span> <span class="va">unique_regions</span><span class="op">[</span><span class="fu">subjectHits</span><span class="op">(</span><span class="va">hits</span><span class="op">)</span><span class="op">]</span>
<span class="co"># and ...</span>
<span class="va">overlap_results</span> <span class="op">&lt;-</span> <span class="va">results</span><span class="op">[</span><span class="fu">queryHits</span><span class="op">(</span><span class="va">hits</span><span class="op">)</span><span class="op">]</span>

<span class="co"># coordinates of overlapping region: contig:start-end:strand</span>
<span class="co"># bin sites in unique_regions</span>
<span class="va">overlap_results</span><span class="op">$</span><span class="va">region</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span>
  <span class="fu">seqnames</span><span class="op">(</span><span class="va">overlap_regions</span><span class="op">)</span>, <span class="st">":"</span>,
  <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="va">overlap_regions</span><span class="op">)</span>, <span class="st">"-"</span>, <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">overlap_regions</span><span class="op">)</span>, <span class="st">":"</span>,
  <span class="fu">strand</span><span class="op">(</span><span class="va">overlap_regions</span><span class="op">)</span>
<span class="op">)</span>

<span class="co"># add site position within extended region</span>
<span class="va">overlap_results</span><span class="op">$</span><span class="va">position</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="va">overlap_results</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="va">overlap_regions</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">overlap_results</span><span class="op">$</span><span class="va">position</span><span class="op">)</span>, xlab <span class="op">=</span> <span class="st">"position of site within a motif\n(original site position is 3)"</span><span class="op">)</span></code></pre></div>
<p><img src="JACUSA2helper-nanopore_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="create-feature-matrix">Create feature matrix<a class="anchor" aria-label="anchor" href="#create-feature-matrix"></a>
</h3>
<p>The pre-processing is almost done, the data frame should contain the following columns:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">mcols</span><span class="op">(</span><span class="va">overlap_results</span><span class="op">)</span>
<span class="co">#&gt; DataFrame with 6365925 rows and 6 columns</span>
<span class="co">#&gt;         meta_cond call2_score insertion_score deletion_score</span>
<span class="co">#&gt;          &lt;factor&gt;   &lt;numeric&gt;       &lt;numeric&gt;      &lt;numeric&gt;</span>
<span class="co">#&gt; 1       WT_vs_IVT    0.182221        0.973777    -0.00084858</span>
<span class="co">#&gt; 2       WT_vs_IVT    0.200339              NA     0.84031855</span>
<span class="co">#&gt; 3       WT_vs_IVT    0.200339              NA     0.84031855</span>
<span class="co">#&gt; 4       WT_vs_IVT    0.279846              NA             NA</span>
<span class="co">#&gt; 5       WT_vs_IVT    0.279846              NA             NA</span>
<span class="co">#&gt; ...           ...         ...             ...            ...</span>
<span class="co">#&gt; 6365921  WT_vs_KO    0.936613       0.0560419      0.0352544</span>
<span class="co">#&gt; 6365922  WT_vs_KO    0.936613       0.0560419      0.0352544</span>
<span class="co">#&gt; 6365923  WT_vs_KO    0.256918       1.0942557      0.2855172</span>
<span class="co">#&gt; 6365924  WT_vs_KO    0.256918       1.0942557      0.2855172</span>
<span class="co">#&gt; 6365925  WT_vs_KO    0.280098       2.7659899             NA</span>
<span class="co">#&gt;                         region  position</span>
<span class="co">#&gt;                    &lt;character&gt; &lt;numeric&gt;</span>
<span class="co">#&gt; 1              1:14409-14413:-         3</span>
<span class="co">#&gt; 2              1:14413-14417:-         3</span>
<span class="co">#&gt; 3              1:14414-14418:-         2</span>
<span class="co">#&gt; 4              1:14413-14417:-         4</span>
<span class="co">#&gt; 5              1:14414-14418:-         3</span>
<span class="co">#&gt; ...                        ...       ...</span>
<span class="co">#&gt; 6365921 X:155943752-15594375..         3</span>
<span class="co">#&gt; 6365922 X:155943753-15594375..         2</span>
<span class="co">#&gt; 6365923 X:155943752-15594375..         4</span>
<span class="co">#&gt; 6365924 X:155943753-15594375..         3</span>
<span class="co">#&gt; 6365925 X:155943759-15594376..         3</span></code></pre></div>
<p>Use <code><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">tidyr::pivot_wider</a></code> to create the feature matrix:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">feature_matrix</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">overlap_results</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider</a></span><span class="op">(</span>
    id_cols <span class="op">=</span> <span class="va">region</span>, 
    names_from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">meta_cond</span>, <span class="va">position</span><span class="op">)</span>, 
    values_from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">call2_score</span>, <span class="va">insertion_score</span>, <span class="va">deletion_score</span><span class="op">)</span>, 
    names_glue <span class="op">=</span> <span class="st">"{meta_cond}_{.value}_{position}"</span>,
    names_sort <span class="op">=</span> <span class="cn">TRUE</span>,
    values_fill <span class="op">=</span> <span class="fl">0.0</span>,
  <span class="op">)</span></code></pre></div>
<div class="section level4">
<h4 id="add-sequence">Add sequence<a class="anchor" aria-label="anchor" href="#add-sequence"></a>
</h4>
<p>In order to add information if a site is contained in a DRACH motif, we need to retrieve sequence information for the overlapping <code>unique_regions</code> of a site.</p>
<p>If you have a custom FASTA sequence, use <code><a href="https://rdrr.io/pkg/Rsamtools/man/FaFile-class.html" class="external-link">Rsamtools::FaFile</a></code> to load the FASTA file. Otherwise, load a genome from via <a href="https://bioconductor.org/packages/release/bioc/html/BSgenome.html" class="external-link">BSgenome</a>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.NCBI.GRCh38"</span><span class="op">)</span>

<span class="co"># retrieve sequence and convert to character vector</span>
<span class="va">feature_matrix</span><span class="op">$</span><span class="va">motif</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BSgenome/man/getSeq-methods.html" class="external-link">getSeq</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.NCBI.GRCh38</span>, <span class="fu">GRanges</span><span class="op">(</span><span class="va">feature_matrix</span><span class="op">$</span><span class="va">region</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># number of top 10 motifs</span>
<span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">feature_matrix</span><span class="op">$</span><span class="va">motif</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; CCAGC GCAGC CCAGG CCACC GCAGG AGAAG CCAAG GGAGG TGAAG GGAAG </span>
<span class="co">#&gt; 30278 29837 23391 23325 21397 20703 20461 20034 19682 17875</span></code></pre></div>
<p>Next, we add an indicator variable if the DRACH motif ([AGT][AG]AC[ACT]) is present:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">feature_matrix</span><span class="op">$</span><span class="va">DRACH</span> <span class="op">&lt;-</span> <span class="fl">0</span>
<span class="va">feature_matrix</span><span class="op">$</span><span class="va">DRACH</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"[AGT][AG]AC[ACT]"</span>, <span class="va">feature_matrix</span><span class="op">$</span><span class="va">motif</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span>

<span class="va">tbl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">feature_matrix</span><span class="op">$</span><span class="va">DRACH</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">tbl</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">recode</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">tbl</span><span class="op">)</span>, <span class="st">"0"</span> <span class="op">=</span> <span class="st">"no"</span>, <span class="st">"1"</span> <span class="op">=</span> <span class="st">"yes"</span><span class="op">)</span>  <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">" ("</span>, <span class="va">tbl</span>, <span class="st">")"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/pie.html" class="external-link">pie</a></span><span class="op">(</span><span class="va">tbl</span>, main<span class="op">=</span><span class="st">"DRACH motif present"</span><span class="op">)</span></code></pre></div>
<p><img src="JACUSA2helper-nanopore_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>Use <code>saveRDS(feature_matrix, file = "feature_matrix.rds")</code> to store the current state. Use <code>feature_matrix &lt;- readRDS("feature_matrix.rds")</code> to restore previous state.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="estimation">Estimation<a class="anchor" aria-label="anchor" href="#estimation"></a>
</h2>
<div class="section level3">
<h3 id="learn-from-miclip-data">Learn from miCLIP Data<a class="anchor" aria-label="anchor" href="#learn-from-miclip-data"></a>
</h3>
<p>Next, we need miCLIP data to train our model: * Boulias, * Koertel, and * Koh. We use sites that are present in all three data sets.</p>
<p>The column “region” corresponds to the extended sites around identified miCLIP sites. Make sure that “region” is 1-indexed if you use your own annotation.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">miclip</span><span class="op">)</span>
<span class="co">#&gt;   contig     start       end     overlap score strand                  region</span>
<span class="co">#&gt; 1      1   1000217   1000222     Boulias     0      -     1:1000218-1000222:-</span>
<span class="co">#&gt; 2      1   1000221   1000226 Koertel,Koh     0      -     1:1000222-1000226:-</span>
<span class="co">#&gt; 3      1 100022414 100022419     Boulias     0      + 1:100022415-100022419:+</span>
<span class="co">#&gt; 4      1 100022490 100022495     Boulias     0      + 1:100022491-100022495:+</span>
<span class="co">#&gt; 5      1   1000246   1000251         Koh     0      -     1:1000247-1000251:-</span>
<span class="co">#&gt; 6      1   1000520   1000525     Koertel     0      -     1:1000521-1000525:-</span></code></pre></div>
<p>Intersect regions of feature matrix with miCLIP data (contained in all three data sets: Boulias,Koertel, and Koh):</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># dimensions of feature matrix BEFORE intersect with miCLIP</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Dimensions of feature matrix"</span><span class="op">)</span>
<span class="co">#&gt; [1] "Dimensions of feature matrix"</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">feature_matrix</span><span class="op">)</span>
<span class="co">#&gt; [1] 2185892      33</span>

<span class="co"># index vector of regions contained in miCLIP data</span>
<span class="va">i</span> <span class="op">&lt;-</span> <span class="va">feature_matrix</span><span class="op">$</span><span class="va">region</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">miclip</span><span class="op">[</span><span class="va">miclip</span><span class="op">$</span><span class="va">overlap</span> <span class="op">==</span> <span class="st">"Boulias,Koertel,Koh"</span>, <span class="st">"region"</span><span class="op">]</span>
<span class="co"># retain sites that overlap with miCLIP and remove non-numeric columns</span>
<span class="va">nmf_matrix</span> <span class="op">&lt;-</span> <span class="va">feature_matrix</span><span class="op">[</span><span class="va">i</span>, <span class="op">]</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># dimensions of feature matrix AFTER intersect with miCLIP</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Dimensions of nmf_matrix after processing"</span><span class="op">)</span>
<span class="co">#&gt; [1] "Dimensions of nmf_matrix after processing"</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">nmf_matrix</span><span class="op">)</span>
<span class="co">#&gt; [1] 2748   33</span></code></pre></div>
<p>The learning algorithm requires the matrix to consist only of numeric value. Remove non-numeric columns:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># dimensions of feature matrix BEFORE intersect with miCLIP</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Dimensions of feature matrix"</span><span class="op">)</span>
<span class="co">#&gt; [1] "Dimensions of feature matrix"</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">nmf_matrix</span><span class="op">)</span>
<span class="co">#&gt; [1] 2748   33</span>

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">nmf_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">nmf_matrix</span><span class="op">$</span><span class="va">region</span>

<span class="co"># remove non-numeric columns</span>
<span class="va">nmf_matrix</span> <span class="op">&lt;-</span> <span class="va">nmf_matrix</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="op">-</span><span class="va">region</span>, <span class="op">-</span><span class="va">motif</span><span class="op">)</span>

<span class="co"># set sensible defaults</span>
<span class="va">nmf_matrix</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">nmf_matrix</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span><span class="fl">0</span>
<span class="va">nmf_matrix</span><span class="op">[</span><span class="va">nmf_matrix</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span>

<span class="co"># dimensions of feature matrix AFTER intersect with miCLIP</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Dimensions of feature matrix after processing"</span><span class="op">)</span>
<span class="co">#&gt; [1] "Dimensions of feature matrix after processing"</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">nmf_matrix</span><span class="op">)</span>
<span class="co">#&gt; [1] 2748   31</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="estimate-parameters">Estimate parameters<a class="anchor" aria-label="anchor" href="#estimate-parameters"></a>
</h3>
<p>Make sure to remove columns such as: “motif” and “region” before you start the learning algorithm.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://renozao.github.io/NMF" class="external-link">NMF</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: pkgmaker</span>
<span class="co">#&gt; Loading required package: registry</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'pkgmaker'</span>
<span class="co">#&gt; The following object is masked from 'package:S4Vectors':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     new2</span>
<span class="co">#&gt; Loading required package: rngtools</span>
<span class="co">#&gt; Loading required package: cluster</span>
<span class="co">#&gt; NMF - BioConductor layer [OK] | Shared memory capabilities [NO: bigmemory] | Cores 31/32</span>
<span class="co">#&gt;   To enable shared memory capabilities, try: install.extras('</span>
<span class="co">#&gt; NMF</span>
<span class="co">#&gt; ')</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'NMF'</span>
<span class="co">#&gt; The following object is masked from 'package:S4Vectors':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     nrun</span>
<span class="fu"><a href="https://rdrr.io/pkg/NMF/man/nmfSeed.html" class="external-link">nmfSeed</a></span><span class="op">(</span><span class="st">'nndsvd'</span><span class="op">)</span>
<span class="co">#&gt; &lt;object of class:  NMFSeed &gt;</span>
<span class="co">#&gt; name:     nndsvd </span>
<span class="co">#&gt; method:   &lt;function&gt;</span>
<span class="va">meth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NMF/man/nmfAlgorithm.html" class="external-link">nmfAlgorithm</a></span><span class="op">(</span>version<span class="op">=</span><span class="st">'R'</span><span class="op">)</span>
<span class="va">meth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">meth</span><span class="op">)</span>, <span class="va">meth</span><span class="op">)</span>

<span class="va">small_nmf_matrix</span> <span class="op">&lt;-</span> <span class="va">nmf_matrix</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">]</span> 

<span class="va">estim.r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NMF/man/nmf.html" class="external-link">nmf</a></span><span class="op">(</span><span class="va">small_nmf_matrix</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">10</span>, nrun <span class="op">=</span> <span class="fl">10</span>, seed <span class="op">=</span> <span class="fl">123456</span>, .opt <span class="op">=</span> <span class="st">'vp3'</span><span class="op">)</span>
<span class="co">#&gt; Compute NMF rank= 2  ... + measures ... OK</span>
<span class="co">#&gt; Compute NMF rank= 3  ... + measures ... OK</span>
<span class="co">#&gt; Compute NMF rank= 4  ... + measures ... OK</span>
<span class="co">#&gt; Compute NMF rank= 5  ... + measures ... OK</span>
<span class="co">#&gt; Compute NMF rank= 6  ... + measures ... OK</span>
<span class="co">#&gt; Compute NMF rank= 7  ... + measures ... OK</span>
<span class="co">#&gt; Compute NMF rank= 8  ... + measures ... OK</span>
<span class="co">#&gt; Compute NMF rank= 9  ... + measures ... OK</span>
<span class="co">#&gt; Compute NMF rank= 10  ... + measures ... OK</span>
<span class="va">V.random</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NMF/man/randomize.html" class="external-link">randomize</a></span><span class="op">(</span><span class="va">small_nmf_matrix</span><span class="op">)</span>

<span class="co"># estimate quality measures from the shuffled data (use default NMF algorithm)</span>
<span class="va">estim.r.random</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NMF/man/nmf.html" class="external-link">nmf</a></span><span class="op">(</span><span class="va">V.random</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">10</span>, nrun<span class="op">=</span><span class="fl">10</span>, seed<span class="op">=</span><span class="fl">123456</span>, .opt<span class="op">=</span><span class="st">'vp3'</span><span class="op">)</span>
<span class="co">#&gt; Compute NMF rank= 2  ... + measures ... OK</span>
<span class="co">#&gt; Compute NMF rank= 3  ... + measures ... OK</span>
<span class="co">#&gt; Compute NMF rank= 4  ... + measures ... OK</span>
<span class="co">#&gt; Compute NMF rank= 5  ... + measures ... OK</span>
<span class="co">#&gt; Compute NMF rank= 6  ... + measures ... OK</span>
<span class="co">#&gt; Compute NMF rank= 7  ... + measures ... OK</span>
<span class="co">#&gt; Compute NMF rank= 8  ... + measures ... OK</span>
<span class="co">#&gt; Compute NMF rank= 9  ... + measures ... OK</span>
<span class="co">#&gt; Compute NMF rank= 10  ... + measures ... OK</span></code></pre></div>
<div class="section level4">
<h4 id="visualize-estimation">Visualize estimation<a class="anchor" aria-label="anchor" href="#visualize-estimation"></a>
</h4>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">estim.r</span>, <span class="va">estim.r.random</span><span class="op">)</span></code></pre></div>
<p><img src="JACUSA2helper-nanopore_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="choose-factorization-rank">Choose factorization rank<a class="anchor" aria-label="anchor" href="#choose-factorization-rank"></a>
</h4>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">DeltaSil</span> <span class="op">&lt;-</span> <span class="va">estim.r</span><span class="op">$</span><span class="va">measures</span><span class="op">$</span><span class="va">silhouette.consensus</span> <span class="op">-</span> <span class="va">estim.r.random</span><span class="op">$</span><span class="va">measures</span><span class="op">$</span><span class="va">silhouette.consensus</span>
<span class="va">DeltaCoph</span> <span class="op">&lt;-</span> <span class="va">estim.r</span><span class="op">$</span><span class="va">measures</span><span class="op">$</span><span class="va">cophenetic</span> <span class="op">-</span> <span class="va">estim.r.random</span><span class="op">$</span><span class="va">measures</span><span class="op">$</span><span class="va">cophenetic</span>

<span class="co">#how can we select factorization rank ?</span>
<span class="va">ChoseRank</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">DeltaSil</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">DeltaSil</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span>,
  <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">DeltaCoph</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">DeltaCoph</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span>
<span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">val_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">DeltaCoph</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="va">val_matrix</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">DeltaSil</span> 
<span class="va">val_matrix</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">DeltaCoph</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">val_matrix</span><span class="op">)</span>, names.arg <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">DeltaCoph</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span>, beside <span class="op">=</span> <span class="cn">TRUE</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009E73"</span>, <span class="st">"#D55E00"</span><span class="op">)</span>, legend.text <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Silhouette"</span>, <span class="st">"Cophenetic"</span><span class="op">)</span>,ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/pretty.html" class="external-link">pretty</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">DeltaSil</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, xlab<span class="op">=</span><span class="st">"Rank"</span>, 
        ylab<span class="op">=</span><span class="st">"Delta"</span>, main <span class="op">=</span> <span class="st">"Difference between original and randomized data"</span>,cex.names<span class="op">=</span><span class="fl">2</span>,cex.axis<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>h<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">DeltaSil</span><span class="op">)</span> , <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">DeltaCoph</span><span class="op">)</span><span class="op">)</span> , col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#009E73"</span>, <span class="st">"#D55E00"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="JACUSA2helper-nanopore_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>final_nmf_matrix <span class="ot">&lt;-</span> <span class="fu">nmf</span>(small_nmf_matrix, ChoseRank, <span class="at">nrun=</span><span class="dv">10</span>, <span class="at">seed=</span><span class="dv">123456</span>, <span class="at">.opt=</span><span class="st">'vp3'</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; NMF algorithm: 'brunet'</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Multiple runs: 10</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Mode: parallel (3/32 core(s))</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>Runs<span class="sc">:</span> <span class="er">|</span>                                                        </span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>Runs<span class="sc">:</span> <span class="er">|</span>                                                  <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>Runs<span class="sc">:</span> <span class="er">|</span>                                                        </span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>Runs<span class="sc">:</span> <span class="er">|==================================================|</span> <span class="dv">100</span>%</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; System time:</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    user  system elapsed </span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  14.469   2.906  10.532</span></span></code></pre></div>
<p>Use <code>saveRDS(final_nmf_matrix, file="final_nmf_matrix.rds")</code> to store final result.</p>
</div>
</div>
<div class="section level3">
<h3 id="visualize-nmf">Visualize NMF<a class="anchor" aria-label="anchor" href="#visualize-nmf"></a>
</h3>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># basis components</span>
<span class="fu"><a href="https://rdrr.io/pkg/NMF/man/heatmaps.html" class="external-link">basismap</a></span><span class="op">(</span><span class="va">final_nmf_matrix</span><span class="op">)</span></code></pre></div>
<p><img src="JACUSA2helper-nanopore_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<p>#<code>{r} mixture coefficients coefmap(final_nmf_matrix) #</code></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">w</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NMF/man/basis-coef-methods.html" class="external-link">basis</a></span><span class="op">(</span><span class="va">final_nmf_matrix</span><span class="op">)</span>
<span class="co"># Table of number of best hits for each pattern"</span>
<span class="va">pattern_instances</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">w</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span> <span class="va">x</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">)</span> <span class="op">}</span> <span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span>
  height <span class="op">=</span> <span class="va">pattern_instances</span>,
  names <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">w</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>,
  main <span class="op">=</span> <span class="st">"NMF_Patterns_Scoring"</span>, 
  xlab <span class="op">=</span> <span class="st">"Patterns"</span>, 
  ylab <span class="op">=</span> <span class="st">"Membership Score (basis matrix)"</span>,
  ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/pretty.html" class="external-link">pretty</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">pattern_instances</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p><img src="JACUSA2helper-nanopore_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<!--
## Visualize patterns

#```{r}
color_palette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
#```

#```{r}
w <- basis(final_nmf_matrix)
h <- coef(final_nmf_matrix)
h <- rbind(h, t(colSums(h)))
#```

#```{r}
## add the combined patterns
#for (pattern in list(c(1,2,4,6), c(1,2,3,4,6))) {
#  h <- rbind(h, t(colSums(h[pattern, ])))
#}
#```

#```{r}
# barplots of patterns
for (k in 1:nrow(h)){
  rep1 <- matrix(h[k, ], ncol = 5, byrow = TRUE)[1:3, ]
  rownames(rep1) <- c("Mismatch","Deletion","Insertion")
  colnames(rep1) <- c("Pos1","Pos2","Pos3","Pos4","Pos5")
  
  barplot(
    rep1,
    main = paste0("NMF_Pattern_", k),
    col = color_palette,
    cex.names = 2,
    cex.axis = 2,
    legend = TRUE
  )
}
```
-->
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Piechotta2021" class="csl-entry">
Piechotta, Michael, Qi Wang, Janine Altmüller, and Christoph Dieterich. 2021. <span>“<span>RNA</span> Modification Mapping with <span>JACUSA</span>2,”</span> July. <a href="https://doi.org/10.1101/2021.07.02.450888" class="external-link">https://doi.org/10.1101/2021.07.02.450888</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Michael Piechotta, <a href="http://dieterichlab.org/" class="external-link">Dieterich Lab</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
