<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="JACUSA2helper">
<title>Introduction to meta conditions with JACUSA2helper • JACUSA2helper</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to meta conditions with JACUSA2helper">
<meta property="og:description" content="JACUSA2helper">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">JACUSA2helper</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.9.9.9500</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/JACUSA2helper.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/JACUSA2helper-dart-seq.html">Analysis of DART-seq with JACUSA2helper</a>
    <a class="dropdown-item" href="../articles/JACUSA2helper-mazf.html">Analysis of m6A mapping by MazF with JACUSA2helper</a>
    <a class="dropdown-item" href="../articles/JACUSA2helper-meta-condtions.html">Introduction to meta conditions with JACUSA2helper</a>
    <a class="dropdown-item" href="../articles/web_only/JACUSA2helper-nanopore.html">Analysis of Nanopore HEK293 with JACUSA2helper</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/dieterich-lab/JACUSA2helper/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Introduction to meta conditions with JACUSA2helper</h1>
                        <h4 data-toc-skip class="author">Michael
Piechotta</h4>
            
            <h4 data-toc-skip class="date">2022-12-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/dieterich-lab/JACUSA2helper/blob/HEAD/vignettes/JACUSA2helper-meta-condtions.Rmd" class="external-link"><code>vignettes/JACUSA2helper-meta-condtions.Rmd</code></a></small>
      <div class="d-none name"><code>JACUSA2helper-meta-condtions.Rmd</code></div>
    </div>

    
    
<p>See <code><a href="../articles/JACUSA2helper.html">vignette("JACUSA2helper")</a></code> for general description of
analysis with JACUSA2helper. For details on JACUSA2, check the <a href="https://github.com/dieterich-lab/JACUSA2/blob/master/manual/manual.pdf" class="external-link">JACUSA2
manual</a>.</p>
<p>In the following, the use of meta conditions to combine multiple
pairwise comparisons will be shown.</p>
<div class="section level3">
<h3 id="meta-condition">Meta condition<a class="anchor" aria-label="anchor" href="#meta-condition"></a>
</h3>
<p>JACUSA2helper supports the analysis of several related JACUSA2 result
files via <code>results &lt;- read_results(files, meta_conds)</code>
where <code>meta_conds</code> is a vector of strings that provides a
descriptive name for each file in the vector of strings
<code>files</code>.</p>
<p>Here, we will use the the <span class="citation">Zhou et al. (<a href="#ref-Zhou2018" role="doc-biblioref">2018</a>)</span> data set,
where the authors map RNA modification of pseudouridine (<span class="math inline">\(\Psi\)</span>) by chemically modifying
pseudouridines with carbodiimide (+CMC) and detecting arrest events that
are induced by reverse transcription stops in high-throughput sequencing
under 3 different conditions (HIVRT, SIIIMn, and SIIIMg). All three data
sets are available in JACUSA2helper via <code><a href="https://rdrr.io/r/utils/data.html" class="external-link">data()</a></code>.
Additionally, we have compiled combined data sets
<code>data(Zhou2018_call2)</code> and
<code>data(Zhou2018_rt_arrest)</code> that utilizes meta conditions.
Note, that this data has NO replicates!</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Zhou2018_rt_arrest</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">Zhou2018_rt_arrest</span><span class="op">$</span><span class="va">meta_cond</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] HIVRT    SIIIRTMn SIIIRTMg</span></span>
<span><span class="co">#&gt; Levels: HIVRT SIIIRTMg SIIIRTMn</span></span></code></pre></div>
<div class="section level4">
<h4 id="group-by-site-and-other">Group By Site (and other)<a class="anchor" aria-label="anchor" href="#group-by-site-and-other"></a>
</h4>
<p>When manipulating a multi <code>results</code> object created by
<code><a href="../reference/read_results.html">read_results()</a></code>, it is crucial to distinguish the following
files in an analysis pipeline:</p>
<ul>
<li><code>results %&gt;% ... other functions()</code></li>
<li><code>results %&gt;% dplyr:group_by(results, meta_cond) %&gt;% ... other functions()</code></li>
</ul>
<p>The first statement will apply any subsequent functions to ALL sites
regardless of the meta condition while the last statement will apply to
sites of EACH meta condition!</p>
</div>
</div>
<div class="section level3">
<h3 id="number-of-sites">Number of sites<a class="anchor" aria-label="anchor" href="#number-of-sites"></a>
</h3>
<p>The following statement will determine the number of covered sites
per contig and meta condition:</p>
<!-- TODO
#```{r}
#GenomicRanges::mcols(Zhou2018_rt_arrest) %>% as.data.frame() %>%
#  dplyr::distinct(contig, meta_cond) %>% 
#  dplyr::group_by(contig, meta_cond) %>% 
#  dplyr::count()
#```
-->
<p>This statement will determine the number of covered sites per contig
regardless of the meta condition:</p>
<!-- TODO
#```{r}
#dplyr::distinct(Zhou2018_rt_arrest, contig, start, end, strand) %>% 
#  dplyr::group_by(contig) %>% 
#  dplyr::count() # count in a group
#```
-->
</div>
<div class="section level2">
<h2 id="filter">Filter<a class="anchor" aria-label="anchor" href="#filter"></a>
</h2>
<p>Filter by coverage regardless of the meta condition.</p>
<!--TODO
#```{r}
#filtered_all <- Zhou2018_rt_arrest %>%
#  dplyr::filter(All(cov$cond1 >= 100) & All(cov$cond2 >= 1000)) %>% 
#  dplyr::group_by(contig, meta_cond) %>% 
#  dplyr::count() # count in a group
#filtered_all
#```
-->
</div>
<div class="section level2">
<h2 id="plot">Plot<a class="anchor" aria-label="anchor" href="#plot"></a>
</h2>
<p>First, we add a description <code>data_desc</code> of the conditions
to the result object. The data sets of <code>Zhou2018</code> have been
layout out in such a way that condition 1 and 2 correspond to
carbodiimide (+CMC) treatment and control (-CMC), respectively.</p>
<!-- TODO
#```{r}
#result <- gather_repl(Zhou2018_rt_arrest$id, Zhou2018_rt_arrest$arrest_rate) %>%
#  dplyr::inner_join(Zhou2018_rt_arrest %>% dplyr::select(id, pvalue, meta_cond), by = "id") %>%
#  dplyr::mutate(
#    data_desc = dplyr::case_when(
#      cond == "cond1" ~ "+CMC",
#      cond == "cond2" ~ "-CMC",
#    )
#  )
#
## grouping enables nice legends in plot
#group <- do.call(interaction, result[c("cond", "repl")])
#result$group <- group
## relate group={cond(ition), repl(icat)}, and nice label
#meta_desc <- dplyr::distinct(result, cond, repl, group, data_desc)
## map group values to nice labels (data_desc)
#limits <- as.vector(meta_desc[["group"]])
#labels <- meta_desc[["data_desc"]]
#```
-->
<p>Next, we define a ggplot2 object that allows to merge legend for
different scales. Check <a href="https://ozh2014.wordpress.com/2017/06/26/combine-related-legends-into-one-in-ggplot2/" class="external-link">combine
legends</a> for details. In brief, we use <code>colour</code> to
represent cond(ition) and <code>linetype</code> to represent repl(icate)
an relate their possible combinations to descriptive
<code>labels</code>.</p>
<!--
#```{r}
#colour_scale <- ggplot2::scale_colour_manual(
#  name = "Treatment",
#  labels = labels,
#  limits = limits,
#  values = factor(meta_desc[["cond"]]) %>% as.integer()
#)
#linetype_scale <- ggplot2::scale_linetype_manual(
#  name = "Treatment",
#  labels = labels,
#  limits = limits,
#  values = factor(meta_desc[["repl"]]) %>% as.integer()
#)
#```
-->
<div class="section level3">
<h3 id="pvalue-distribution">Pvalue distribution<a class="anchor" aria-label="anchor" href="#pvalue-distribution"></a>
</h3>
<p>Plot of the empirical cumulative distributions of pvalues for all
meta conditions.</p>
<!--
#```{r}
result %>% 
  ggplot2::ggplot(ggplot2::aes(x = pvalue, colour = meta_cond)) +
  ggplot2::labs(colour = "Meta condition") +
  ggplot2::ylab("Density") +
  ggplot2::stat_ecdf(geom = "step") +
  ggplot2::xlab("pvalue") +
  ggplot2::theme(legend.position = "bottom") 
```


## Empirical cumulative arrest rate distribution
#```{r}
result %>% 
  ggplot2::ggplot(ggplot2::aes(x = value, colour = group, linetype = group)) +
  ggplot2::stat_ecdf(geom = "step") +
  colour_scale +
  linetype_scale +
  ggplot2::xlab("Density") +
  ggplot2::xlab("Arrest rate") +
  ggplot2::theme(legend.position = "bottom") + 
  ggplot2::facet_grid(meta_cond ~ .)
```

## Compare coverage
#```{r}
gather_repl(Zhou2018_rt_arrest$id, Zhou2018_rt_arrest$cov) %>%
  dplyr::inner_join(dplyr::select(Zhou2018_rt_arrest, id, meta_cond), by = "id") %>%
  dplyr::mutate(
    data_desc = dplyr::case_when(
      cond == "cond1" ~ "+CMC",
      cond == "cond2" ~ "-CMC",
    )
  ) %>%
  ggplot2::ggplot(ggplot2::aes(x = value, colour = group, linetype = group)) +
  ggplot2::stat_ecdf(geom = "step") +
  colour_scale +
  linetype_scale +
  ggplot2::scale_x_log10(
    breaks = scales::trans_breaks("log10", function(x) 10^x),
    labels = scales::trans_format("log10", scales::math_format(10^.x))
  ) + 
  ggplot2::xlab("Read Coverage") +
  ggplot2::theme(legend.position = "bottom") + 
  ggplot2::facet_grid(meta_cond ~ .)
#```


## Compare arrest rate

Volcano plot of different library types (ignoring 0 arrest rate sites):

#```{r}
Zhou2018_rt_arrest %>%
  dplyr::filter(strand == "+") %>%
  ggplot2::ggplot(
    ggplot2::aes(
      x = log2((arrest_rate$cond1$rep1) / (arrest_rate$cond2$rep1)),
      y = -log10(pvalue),
      colour = meta_cond
    )
  ) +
  ggplot2::geom_point() + 
  ggplot2::geom_hline(yintercept = 2, linetype = 3, colour = "red") +
  ggplot2::ylab("-log10(pvalue)") + 
  ggplot2::xlab("log2(arrest rate +GMC / -GMC)") + 
  ggplot2::facet_grid(. ~ meta_cond)
  
#```
We deduce that `p <= 0.01` is a sensible cutoff.
We plot the arrest rate for all three meta conditions and +CMC and -CMC treatment.

#```{r}
Zhou2018_rt_arrest %>%
  dplyr::filter(strand == "+") %>%
  dplyr::filter(pvalue <= 0.01) %>%
  dplyr::group_by(meta_cond) %>%
  ggplot2::ggplot(ggplot2::aes(x = log2(arrest_rate$cond1$rep1), y = log2(arrest_rate$cond2$rep1), colour = meta_cond)) +
    ggplot2::geom_point() + 
    ggplot2::ylab("log2(Arrest rate -CMC)") + 
    ggplot2::xlab("log2(Arrest rate +CMC)") + 
    ggplot2::facet_grid(. ~ meta_cond) +
    ggplot2::geom_abline(linetype = "dashed", colour = "gray") +
    ggplot2::labs(colour = "Meta condition") +
    ggplot2::theme(legend.position = "bottom")
#```

We plot the arrest rate delta of +CMC and -CMC treatment for all three meta 
conditions and add Pseudouridine sites (black points).

#```{r}
#data(rnam)
#retain_rrna <- "18S"
#filtered_rnam <- rnam %>% 
#  dplyr::filter(rrna == retain_rrna, base == "psU")
#
#Zhou2018_rt_arrest %>%
#  dplyr::filter(contig == retain_rrna) %>%
#  dplyr::filter(strand == "+") %>%
#  dplyr::filter(pvalue <= 0.01) %>%
#  dplyr::group_by(meta_cond) %>%
#  dplyr::mutate(arrest_rate_delta = arrest_rate$cond1$rep1 - arrest_rate$cond2$rep1) %>%
#  ggplot2::ggplot(ggplot2::aes(x = end, y = arrest_rate_delta, fill = meta_cond)) +
#    ggplot2::geom_bar(stat = "identity") +
#    ggplot2::geom_point(data = filtered_rnam, ggplot2::aes(x = pos, y = 0), fill = "black") +
#    ggplot2::xlab(paste0("Position on ", retain_rrna)) +
#    ggplot2::ylab("Arrest rate (+CMC - -CMC)") +
#    ggplot2::ylim(c(-0.1, 0.1)) +
#    ggplot2::theme(legend.position = "bottom") +
#    ggplot2::labs(fill = "Meta condition")
#```
-->
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Zhou2018" class="csl-entry">
Zhou, Katherine I., Wesley C. Clark, David W. Pan, Matthew J. Eckwahl,
Qing Dai, and Tao Pan. 2018. <span>“Pseudouridines Have
Context-Dependent Mutation and Stop Rates in High-Throughput
Sequencing.”</span> <em><span>RNA</span> Biology</em> 15 (7): 892–900.
<a href="https://doi.org/10.1080/15476286.2018.1462654" class="external-link">https://doi.org/10.1080/15476286.2018.1462654</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Michael Piechotta, <a href="http://dieterichlab.org/" class="external-link">Dieterich Lab</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
